{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "14e44b13-45db-4e76-a95c-117b16104e38",
   "metadata": {},
   "source": [
    "# üß¨ Example Report: Genes to SNPs\n",
    "This notebook demonstrates how to run the Gene ‚Üí SNP report using Biofilter3R.\n",
    "It shows how to:\n",
    "- Check database connection and schema version\n",
    "- List available reports\n",
    "- View report documentation\n",
    "- Run the report using hardcoded gene lists\n",
    "- Run the report from a file-based list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5758400f-61b9-487f-9723-b59aae017b0c",
   "metadata": {},
   "source": [
    "### üì¶ Step 1: Import Biofilter and connect to the database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4be6574c-8d10-4676-a1ec-f8b355ceade1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from biofilter import Biofilter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "aa2f193c-eea0-4fac-87be-f9e5c7eb1a1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace with the appropriate connection string for your environment\n",
    "db_uri = \"postgresql+psycopg2://bioadmin:bioadmin@localhost/biofilter\"\n",
    "\n",
    "# Instance of Biofilter\n",
    "bf = Biofilter(db_uri)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c9d1eb95-fd5a-4f1e-a04c-3da3643435e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Input Gene</th>\n",
       "      <th>HGNC Symbol</th>\n",
       "      <th>Matched Name</th>\n",
       "      <th>Alias Type</th>\n",
       "      <th>Alias Source</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Variant ID</th>\n",
       "      <th>Variant Type</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref Allele</th>\n",
       "      <th>Alt Allele</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>246126</td>\n",
       "      <td>TXLNGY</td>\n",
       "      <td>246126</td>\n",
       "      <td>code</td>\n",
       "      <td>ENTREZ</td>\n",
       "      <td>38610</td>\n",
       "      <td>rs3900</td>\n",
       "      <td>SNV</td>\n",
       "      <td>Y</td>\n",
       "      <td>19568371.0</td>\n",
       "      <td>19568371.0</td>\n",
       "      <td>[\"G\"]</td>\n",
       "      <td>[\"C\"]</td>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>GRCh38.p14</td>\n",
       "      <td>8.9</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>246126</td>\n",
       "      <td>TXLNGY</td>\n",
       "      <td>246126</td>\n",
       "      <td>code</td>\n",
       "      <td>ENTREZ</td>\n",
       "      <td>38610</td>\n",
       "      <td>rs3902</td>\n",
       "      <td>SNV</td>\n",
       "      <td>Y</td>\n",
       "      <td>19568761.0</td>\n",
       "      <td>19568761.0</td>\n",
       "      <td>[\"A\"]</td>\n",
       "      <td>[\"G\"]</td>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>GRCh38.p14</td>\n",
       "      <td>8.3</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>246126</td>\n",
       "      <td>TXLNGY</td>\n",
       "      <td>246126</td>\n",
       "      <td>code</td>\n",
       "      <td>ENTREZ</td>\n",
       "      <td>38610</td>\n",
       "      <td>rs3908</td>\n",
       "      <td>DELINS</td>\n",
       "      <td>Y</td>\n",
       "      <td>19571282.0</td>\n",
       "      <td>19571282.0</td>\n",
       "      <td>[\"GGGG\"]</td>\n",
       "      <td>[\"GGG\"]</td>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>GRCh38.p14</td>\n",
       "      <td>8.1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>246126</td>\n",
       "      <td>TXLNGY</td>\n",
       "      <td>246126</td>\n",
       "      <td>code</td>\n",
       "      <td>ENTREZ</td>\n",
       "      <td>38610</td>\n",
       "      <td>rs3909</td>\n",
       "      <td>DELINS</td>\n",
       "      <td>Y</td>\n",
       "      <td>19571277.0</td>\n",
       "      <td>19571278.0</td>\n",
       "      <td>[\"A\"]</td>\n",
       "      <td>[\"AA\"]</td>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>GRCh38.p14</td>\n",
       "      <td>7.5</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>246126</td>\n",
       "      <td>TXLNGY</td>\n",
       "      <td>246126</td>\n",
       "      <td>code</td>\n",
       "      <td>ENTREZ</td>\n",
       "      <td>38610</td>\n",
       "      <td>rs3910</td>\n",
       "      <td>SNV</td>\n",
       "      <td>Y</td>\n",
       "      <td>19571345.0</td>\n",
       "      <td>19571345.0</td>\n",
       "      <td>[\"T\"]</td>\n",
       "      <td>[\"A\"]</td>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>GRCh38.p14</td>\n",
       "      <td>7.5</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>246126</td>\n",
       "      <td>TXLNGY</td>\n",
       "      <td>246126</td>\n",
       "      <td>code</td>\n",
       "      <td>ENTREZ</td>\n",
       "      <td>38610</td>\n",
       "      <td>rs3911</td>\n",
       "      <td>SNV</td>\n",
       "      <td>Y</td>\n",
       "      <td>19571568.0</td>\n",
       "      <td>19571568.0</td>\n",
       "      <td>[\"A\"]</td>\n",
       "      <td>[\"G\"]</td>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>GRCh38.p14</td>\n",
       "      <td>7.8</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>246126</td>\n",
       "      <td>TXLNGY</td>\n",
       "      <td>246126</td>\n",
       "      <td>code</td>\n",
       "      <td>ENTREZ</td>\n",
       "      <td>38610</td>\n",
       "      <td>rs3912</td>\n",
       "      <td>SNV</td>\n",
       "      <td>Y</td>\n",
       "      <td>19570453.0</td>\n",
       "      <td>19570453.0</td>\n",
       "      <td>[\"A\"]</td>\n",
       "      <td>[\"T\"]</td>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>GRCh38.p14</td>\n",
       "      <td>7.5</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>246126</td>\n",
       "      <td>TXLNGY</td>\n",
       "      <td>246126</td>\n",
       "      <td>code</td>\n",
       "      <td>ENTREZ</td>\n",
       "      <td>38610</td>\n",
       "      <td>rs1058414</td>\n",
       "      <td>SNV</td>\n",
       "      <td>Y</td>\n",
       "      <td>19590151.0</td>\n",
       "      <td>19590151.0</td>\n",
       "      <td>[\"G\"]</td>\n",
       "      <td>[\"A\"]</td>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>GRCh38.p14</td>\n",
       "      <td>7.5</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>246126</td>\n",
       "      <td>TXLNGY</td>\n",
       "      <td>246126</td>\n",
       "      <td>code</td>\n",
       "      <td>ENTREZ</td>\n",
       "      <td>38610</td>\n",
       "      <td>rs1179186</td>\n",
       "      <td>SNV</td>\n",
       "      <td>Y</td>\n",
       "      <td>19583386.0</td>\n",
       "      <td>19583386.0</td>\n",
       "      <td>[\"C\"]</td>\n",
       "      <td>[\"A\"]</td>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>GRCh38.p14</td>\n",
       "      <td>6.5</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>246126</td>\n",
       "      <td>TXLNGY</td>\n",
       "      <td>246126</td>\n",
       "      <td>code</td>\n",
       "      <td>ENTREZ</td>\n",
       "      <td>38610</td>\n",
       "      <td>rs1179187</td>\n",
       "      <td>SNV</td>\n",
       "      <td>Y</td>\n",
       "      <td>19582208.0</td>\n",
       "      <td>19582208.0</td>\n",
       "      <td>[\"C\"]</td>\n",
       "      <td>[\"A\", \"G\"]</td>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>GRCh38.p14</td>\n",
       "      <td>7.5</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Input Gene HGNC Symbol Matched Name Alias Type Alias Source  Gene ID  \\\n",
       "0     246126      TXLNGY       246126       code       ENTREZ    38610   \n",
       "1     246126      TXLNGY       246126       code       ENTREZ    38610   \n",
       "2     246126      TXLNGY       246126       code       ENTREZ    38610   \n",
       "3     246126      TXLNGY       246126       code       ENTREZ    38610   \n",
       "4     246126      TXLNGY       246126       code       ENTREZ    38610   \n",
       "5     246126      TXLNGY       246126       code       ENTREZ    38610   \n",
       "6     246126      TXLNGY       246126       code       ENTREZ    38610   \n",
       "7     246126      TXLNGY       246126       code       ENTREZ    38610   \n",
       "8     246126      TXLNGY       246126       code       ENTREZ    38610   \n",
       "9     246126      TXLNGY       246126       code       ENTREZ    38610   \n",
       "\n",
       "  Variant ID Variant Type Chr       Start         End Ref Allele  Alt Allele  \\\n",
       "0     rs3900          SNV   Y  19568371.0  19568371.0      [\"G\"]       [\"C\"]   \n",
       "1     rs3902          SNV   Y  19568761.0  19568761.0      [\"A\"]       [\"G\"]   \n",
       "2     rs3908       DELINS   Y  19571282.0  19571282.0   [\"GGGG\"]     [\"GGG\"]   \n",
       "3     rs3909       DELINS   Y  19571277.0  19571278.0      [\"A\"]      [\"AA\"]   \n",
       "4     rs3910          SNV   Y  19571345.0  19571345.0      [\"T\"]       [\"A\"]   \n",
       "5     rs3911          SNV   Y  19571568.0  19571568.0      [\"A\"]       [\"G\"]   \n",
       "6     rs3912          SNV   Y  19570453.0  19570453.0      [\"A\"]       [\"T\"]   \n",
       "7  rs1058414          SNV   Y  19590151.0  19590151.0      [\"G\"]       [\"A\"]   \n",
       "8  rs1179186          SNV   Y  19583386.0  19583386.0      [\"C\"]       [\"A\"]   \n",
       "9  rs1179187          SNV   Y  19582208.0  19582208.0      [\"C\"]  [\"A\", \"G\"]   \n",
       "\n",
       "      Accession    Assembly  Quality  Note  \n",
       "0  NC_000024.10  GRCh38.p14      8.9  None  \n",
       "1  NC_000024.10  GRCh38.p14      8.3  None  \n",
       "2  NC_000024.10  GRCh38.p14      8.1  None  \n",
       "3  NC_000024.10  GRCh38.p14      7.5  None  \n",
       "4  NC_000024.10  GRCh38.p14      7.5  None  \n",
       "5  NC_000024.10  GRCh38.p14      7.8  None  \n",
       "6  NC_000024.10  GRCh38.p14      7.5  None  \n",
       "7  NC_000024.10  GRCh38.p14      7.5  None  \n",
       "8  NC_000024.10  GRCh38.p14      6.5  None  \n",
       "9  NC_000024.10  GRCh38.p14      7.5  None  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result = bf.report.run_example_report(\"report_gene_to_snp\")\n",
    "result.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d90b4c9-071a-4e25-935b-28ae68800710",
   "metadata": {},
   "source": [
    "### üîç Step 2: Confirm database schema version (optional)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c6d84ad-d693-433c-9281-2250f2eb3c30",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Current Biofilter Schema Version:\", bf.metadata.schema_version)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5a94de8-9f2d-4f89-883b-5935cf3fec5c",
   "metadata": {},
   "source": [
    "### üìÉ Step 3: Explore available reports (optional)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac11b1e3-143b-4b2c-8b62-f77c882ebbcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List all registered reports with descriptions\n",
    "bf.report.list_reports()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f95e018-755d-43b1-87da-41a7c210432a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show detailed explanation for the Gene-to-SNP report\n",
    "bf.report.explain(\"report_gene_to_snp\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ef1a905-95b2-4dc3-8d1a-3a9b0518d944",
   "metadata": {},
   "source": [
    "### ‚ñ∂Ô∏è Step 4: Run the example report (predefined inputs inside the report class)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1300c3f8-534f-40c8-82a2-d40a2d520849",
   "metadata": {},
   "outputs": [],
   "source": [
    "result = bf.report.run_example_report(\"report_gene_to_snp\")\n",
    "\n",
    "# View result\n",
    "result.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82f01970-b8c2-4d04-80dc-f22e106f1216",
   "metadata": {},
   "source": [
    "### üß™ Step 5: Run the report with a custom gene list (inline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e502508-7776-4be9-8f8b-4fa847d04487",
   "metadata": {},
   "outputs": [],
   "source": [
    "custom_gene_list = [\n",
    "    \"TXLNGY\",             # Symbol\n",
    "    \"HGNC:18473\",         # HGNC ID\n",
    "    \"246126\",             # Entrez ID\n",
    "    \"ENSG00000131002\",    # Ensembl ID\n",
    "    \"HGNC:5\"              # Another HGNC ID\n",
    "]\n",
    "\n",
    "result2 = bf.report.run_report(\n",
    "    \"report_gene_to_snp\",\n",
    "    assembly='38',\n",
    "    input_data=custom_gene_list,\n",
    ")\n",
    "\n",
    "# Preview the result\n",
    "result2.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ada48e59-e2ce-49f3-b5e6-87b04bebd69a",
   "metadata": {},
   "source": [
    "### üìÇ Step 6: Run the report using a gene list from a local file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a67e52d9-3034-4931-a8bd-7a4722884361",
   "metadata": {},
   "outputs": [],
   "source": [
    "# File must contain one gene per line (symbol, ID, or alias)\n",
    "result3 = bf.report.run_report(\n",
    "    \"report_gene_to_snp\",\n",
    "    input_data=\"/home/bioadmin/biofilter/notebooks/example_reports/gene_list.txt\"\n",
    ")\n",
    "\n",
    "# Preview result\n",
    "print(len(result3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da4b2fdf-f378-4970-9a1e-a1fd2e04f4f2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
