{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0fe43894-4856-4548-bda8-1bc67dca8e8a",
   "metadata": {},
   "source": [
    "<h1>ðŸ“˜ Biofilter Queries â€” Quick Reference Guide</h1>\n",
    "\n",
    "The Query class provides an interactive, notebook-friendly interface for exploring all database content using clean, Pythonic SQLAlchemy statements. It allows you to build and execute SQL queries, inspect models, and retrieve results as pandas DataFrames with minimal boilerplate.\n",
    "\n",
    "Methods:\n",
    "* ..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e66c31f-bc17-4533-b952-713d267b6e59",
   "metadata": {},
   "source": [
    "<h4>Models vs. Tables</h4>\n",
    "\n",
    "In Biofilter3R, a model is a Python ORM class that represents a biological concept (such as a gene, variant, or disease) and defines its columns, relationships, and behavior inside the system. A table, on the other hand, is the physical structure stored in the database where the actual data lives. Every model maps to one table, but models provide additional structureâ€”such as relationships, typing, and a user-friendly APIâ€”while tables provide efficient storage and indexing. In practice, you query models when working in Python (e.g., select(GeneMaster)), and you reference tables when executing raw SQL (SELECT * FROM gene_master). The distinction allows Biofilter3R to offer both a high-level, intuitive interface for exploration and a low-level interface for performance and debugging."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9d313a7-d680-400f-a4af-6c376580c717",
   "metadata": {},
   "source": [
    "<h4>Query vs. Report</h4>\n",
    "In Biofilter3R, the Query interface provides low-level, flexible access to the database, allowing users to explore \n",
    "tables and models interactively using SQLAlchemy expressions or raw SQL. It is designed for ad-hoc analysis, \n",
    "data inspection, and exploratory work inside notebooks. In contrast, a Report is a high-level, structured analysis module that encapsulates a complete workflowâ€”from parameter validation to querying, post-processing, and formatted output. Reports are reusable, versioned, and intended for standardized scientific or operational tasks, whereas Queries are free-form, exploratory, and user-driven."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb429112-0e04-4df4-abf3-9ff5f06fef1f",
   "metadata": {},
   "source": [
    "---------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8d311e3-b010-419d-840f-3bf68cfc342d",
   "metadata": {},
   "source": [
    "### Start Biofilter3R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d58ce367-5ddb-43f0-a86d-44289aca2e23",
   "metadata": {},
   "outputs": [],
   "source": [
    "from biofilter import Biofilter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1f8471f0-d025-4076-974f-28ecea5b7210",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[INFO] â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•\n",
      "[INFO] ðŸš€ Initializing Biofilter3R\n",
      "[INFO]    â€¢ Version: 3.2.0\n",
      "[INFO]    â€¢ Debug mode: False\n",
      "[INFO]    â€¢ Config: /home/bioadmin/biofilter/.biofilter.toml\n",
      "[INFO] â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•\n",
      "[INFO] ðŸ”Œ Database connection established\n",
      "[INFO]    â€¢ Engine: postgresql+psycopg2\n",
      "[INFO]    â€¢ Host:   localhost\n",
      "[INFO]    â€¢ DB:     biofilter\n",
      "[INFO]    â€¢ Time:   0.9 ms\n",
      "[INFO] â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•\n"
     ]
    }
   ],
   "source": [
    "# Instance of Biofilter\n",
    "bf = Biofilter()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2770a24-434f-4dc8-9cfd-055e5f1b135d",
   "metadata": {},
   "source": [
    "--------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4bff956-257b-4131-9dc8-1b7bc75c8281",
   "metadata": {},
   "source": [
    "### 1. Access a model by name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3139db1f-d2e2-46a0-a3e1-7f412bb55e48",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "biofilter.db.models.model_genes.GeneMaster"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bf.query.GeneMaster"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24b1ff2d-5894-455c-bbe5-f5725c824e26",
   "metadata": {},
   "source": [
    "------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9746573-53fc-45f0-a0a8-5b40170d54b9",
   "metadata": {},
   "source": [
    "### 2. Query a model using simple filters\n",
    "Executes an equality-based query on the chosen model using keyword arguments (e.g., get(\"GeneMaster\", hgnc_id=\"5\")). Automatically resolves the model name, builds the SQL query, and returns the results as a pandas DataFrame. Useful for quick lookups, filtering by one or more columns, and performing fast exploratory queries without writing full SQLAlchemy statements.\n",
    "\n",
    "get() will:\n",
    "* Resolve model name strings\n",
    "* Accept only scalar equality filters\n",
    "* Execute a simple SELECT * FROM model WHERE ...\n",
    "* Return a clean DataFrame\n",
    "\n",
    "query_model will not:\n",
    "* Handle lists (IN (...))\n",
    "* Handle ranges (>, <, BETWEEN)\n",
    "* Handle LIKE, ILIKE, regex\n",
    "* Perform joins between tables\n",
    "* Return ORM objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4d4225b3-77f1-438a-8343-7eeb10157b78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>hgnc_status</th>\n",
       "      <th>omic_status_id</th>\n",
       "      <th>entity_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>data_source_id</th>\n",
       "      <th>etl_package_id</th>\n",
       "      <th>locus_group_id</th>\n",
       "      <th>locus_type_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id symbol hgnc_status  omic_status_id  entity_id chromosome  \\\n",
       "0   1   A1BG    Approved               1          1         19   \n",
       "\n",
       "   data_source_id  etl_package_id  locus_group_id  locus_type_id  \n",
       "0               2               6               1              1  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = bf.query.get(\"GeneMaster\", symbol=\"A1BG\") # use string\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "456ea994-90d6-44f3-bdc6-067ec93682e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position_37</th>\n",
       "      <th>position_38</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>alternate_allele</th>\n",
       "      <th>data_source_id</th>\n",
       "      <th>etl_package_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>171</td>\n",
       "      <td>1</td>\n",
       "      <td>175261679</td>\n",
       "      <td>175292543</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   rs_id  chromosome  position_37  position_38 reference_allele  \\\n",
       "0    171           1    175261679    175292543                T   \n",
       "\n",
       "  alternate_allele  data_source_id  etl_package_id  \n",
       "0                C               4              70  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = bf.query.get(\n",
    "    bf.query.VariantSNP,  # use class object\n",
    "    chromosome=\"1\",\n",
    "    position_38=175292543\n",
    ")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46a9466f-1b5f-4a4b-94a1-875408047bc3",
   "metadata": {},
   "source": [
    "-------------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "264b4d3c-b1cf-40e7-8ec1-dcf5de2c4932",
   "metadata": {},
   "source": [
    "### 3. Query a model using complex statements\n",
    "\n",
    "The model() method executes a SQLAlchemy statement (or model class) and always returns a pandas DataFrame.\n",
    "It is designed for interactive exploration and quick data inspection in python scripts."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "855098b0-039e-4c37-a74f-c0c183dd09dc",
   "metadata": {},
   "source": [
    "#### 3.1. Load all records from a model\n",
    "\n",
    "Passing a model class directly is equivalent to running SELECT * FROM table.\n",
    "This is the fastest way to preview the contents of a table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5e57fc16-51e5-48ba-a46d-8f976345c42a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>hgnc_status</th>\n",
       "      <th>omic_status_id</th>\n",
       "      <th>entity_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>data_source_id</th>\n",
       "      <th>etl_package_id</th>\n",
       "      <th>locus_group_id</th>\n",
       "      <th>locus_type_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>A2M</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id    symbol hgnc_status  omic_status_id  entity_id chromosome  \\\n",
       "0   1      A1BG    Approved               1          1         19   \n",
       "1   2  A1BG-AS1    Approved               1          2         19   \n",
       "2   3      A1CF    Approved               1          3         10   \n",
       "3   4       A2M    Approved               1          4         12   \n",
       "4   5   A2M-AS1    Approved               1          5         12   \n",
       "\n",
       "   data_source_id  etl_package_id  locus_group_id  locus_type_id  \n",
       "0               2               6               1              1  \n",
       "1               2               6               2              2  \n",
       "2               2               6               1              1  \n",
       "3               2               6               1              1  \n",
       "4               2               6               2              2  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = bf.query.run_model(bf.query.GeneMaster)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b40c0daa-622b-4b74-8f65-69678ad8c23e",
   "metadata": {},
   "source": [
    "#### 3.2 Limit the number of rows\n",
    "Use select() when you want to control limits, filters, or joins explicitly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "169be766-7fa6-4dea-8418-83a28e4a5265",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>hgnc_status</th>\n",
       "      <th>omic_status_id</th>\n",
       "      <th>entity_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>data_source_id</th>\n",
       "      <th>etl_package_id</th>\n",
       "      <th>locus_group_id</th>\n",
       "      <th>locus_type_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>A2M</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>A2ML1-AS1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>A2ML1-AS2</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id     symbol hgnc_status  omic_status_id  entity_id chromosome  \\\n",
       "0   1       A1BG    Approved               1          1         19   \n",
       "1   2   A1BG-AS1    Approved               1          2         19   \n",
       "2   3       A1CF    Approved               1          3         10   \n",
       "3   4        A2M    Approved               1          4         12   \n",
       "4   5    A2M-AS1    Approved               1          5         12   \n",
       "5   6      A2ML1    Approved               1          6         12   \n",
       "6   7  A2ML1-AS1    Approved               1          7         12   \n",
       "7   8  A2ML1-AS2    Approved               1          8         12   \n",
       "8   9    A3GALT2    Approved               1          9          1   \n",
       "9  10     A4GALT    Approved               1         10         22   \n",
       "\n",
       "   data_source_id  etl_package_id  locus_group_id  locus_type_id  \n",
       "0               2               6               1              1  \n",
       "1               2               6               2              2  \n",
       "2               2               6               1              1  \n",
       "3               2               6               1              1  \n",
       "4               2               6               2              2  \n",
       "5               2               6               1              1  \n",
       "6               2               6               2              2  \n",
       "7               2               6               2              2  \n",
       "8               2               6               1              1  \n",
       "9               2               6               1              1  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "select = bf.query.select # Select Method\n",
    "genes = bf.query.GeneMaster # GeneMaster Model Class = Table\n",
    "\n",
    "stmt = select(genes).limit(10)\n",
    "\n",
    "df = bf.query.run_model(stmt)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "701ab09e-5d0f-4f56-920e-79301350f6c0",
   "metadata": {},
   "source": [
    "#### 3.3. Filter using a WHERE clause\n",
    "This example retrieves a specific gene using an equality-based filter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "30e07d58-5fcc-4173-90d1-c52afae0f53d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>hgnc_status</th>\n",
       "      <th>omic_status_id</th>\n",
       "      <th>entity_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>data_source_id</th>\n",
       "      <th>etl_package_id</th>\n",
       "      <th>locus_group_id</th>\n",
       "      <th>locus_type_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id symbol hgnc_status  omic_status_id  entity_id chromosome  \\\n",
       "0   1   A1BG    Approved               1          1         19   \n",
       "\n",
       "   data_source_id  etl_package_id  locus_group_id  locus_type_id  \n",
       "0               2               6               1              1  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Used objects from step 2.2\n",
    "stmt = (select(genes).where(genes.symbol == \"A1BG\"))\n",
    "df = bf.query.run_model(stmt)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbdb6a5c-ddc8-4434-a43b-bcfb900d17dd",
   "metadata": {},
   "source": [
    "#### 3.4. Query variants by chromosome and position\n",
    "Multiple where() clauses can be chained to build precise queries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1dcf115a-fcee-4080-8220-d5ed39fbf793",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position_37</th>\n",
       "      <th>position_38</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>alternate_allele</th>\n",
       "      <th>data_source_id</th>\n",
       "      <th>etl_package_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>672</td>\n",
       "      <td>1</td>\n",
       "      <td>53679329</td>\n",
       "      <td>53213657</td>\n",
       "      <td>G</td>\n",
       "      <td>A/C</td>\n",
       "      <td>4</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   rs_id  chromosome  position_37  position_38 reference_allele  \\\n",
       "0    672           1     53679329     53213657                G   \n",
       "\n",
       "  alternate_allele  data_source_id  etl_package_id  \n",
       "0              A/C               4              70  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VariantSNP = bf.query.VariantSNP\n",
    "select = bf.query.select\n",
    "\n",
    "stmt = (\n",
    "    select(VariantSNP)\n",
    "    .where(VariantSNP.chromosome == \"1\")\n",
    "    .where(VariantSNP.position_38 == 53213657)\n",
    ")\n",
    "\n",
    "df = bf.query.run_model(stmt)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac24ba07-bf88-4667-9955-061de3c2da00",
   "metadata": {},
   "source": [
    "#### 3.5. Query multiple records using IN\n",
    "Use .in_() to retrieve multiple records in a single query."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e7c51cf3-3bb8-4a05-968f-331d7103005e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position_37</th>\n",
       "      <th>position_38</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>alternate_allele</th>\n",
       "      <th>data_source_id</th>\n",
       "      <th>etl_package_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>672</td>\n",
       "      <td>1</td>\n",
       "      <td>53679329</td>\n",
       "      <td>53213657</td>\n",
       "      <td>G</td>\n",
       "      <td>A/C</td>\n",
       "      <td>4</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2525600908</td>\n",
       "      <td>1</td>\n",
       "      <td>53680227</td>\n",
       "      <td>53214555</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        rs_id  chromosome  position_37  position_38 reference_allele  \\\n",
       "0         672           1     53679329     53213657                G   \n",
       "1  2525600908           1     53680227     53214555                G   \n",
       "\n",
       "  alternate_allele  data_source_id  etl_package_id  \n",
       "0              A/C               4              70  \n",
       "1                T               4              70  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Used objects from step 2.4\n",
    "\n",
    "positions = [53213657, 53214012, 53214555]\n",
    "\n",
    "stmt = (\n",
    "    select(VariantSNP)\n",
    "    .where(VariantSNP.chromosome == \"1\")\n",
    "    .where(VariantSNP.position_38.in_(positions))\n",
    ")\n",
    "\n",
    "df = bf.query.run_model(stmt)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "468f5394-bc77-453f-a861-22fbfee183b7",
   "metadata": {},
   "source": [
    "#### 3.6. Select specific columns\n",
    "Selecting explicit columns reduces memory usage and improves readability when you only need a subset of fields."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8970a7e4-b52c-4e88-b8fd-07dad4054703",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>hgnc_status</th>\n",
       "      <th>entity_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>Approved</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>A2M</td>\n",
       "      <td>Approved</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72524</th>\n",
       "      <td>72525</td>\n",
       "      <td>MVCD5</td>\n",
       "      <td>Gene from NCBI</td>\n",
       "      <td>88648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72525</th>\n",
       "      <td>72526</td>\n",
       "      <td>DEL2P16.3</td>\n",
       "      <td>Gene from NCBI</td>\n",
       "      <td>88650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72526</th>\n",
       "      <td>72527</td>\n",
       "      <td>PHN</td>\n",
       "      <td>Gene from NCBI</td>\n",
       "      <td>88653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72527</th>\n",
       "      <td>72528</td>\n",
       "      <td>LOC143923313</td>\n",
       "      <td>Gene from NCBI</td>\n",
       "      <td>88655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72528</th>\n",
       "      <td>72529</td>\n",
       "      <td>FL2</td>\n",
       "      <td>Gene from NCBI</td>\n",
       "      <td>88658</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72529 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          id        symbol     hgnc_status  entity_id\n",
       "0          1          A1BG        Approved          1\n",
       "1          2      A1BG-AS1        Approved          2\n",
       "2          3          A1CF        Approved          3\n",
       "3          4           A2M        Approved          4\n",
       "4          5       A2M-AS1        Approved          5\n",
       "...      ...           ...             ...        ...\n",
       "72524  72525         MVCD5  Gene from NCBI      88648\n",
       "72525  72526     DEL2P16.3  Gene from NCBI      88650\n",
       "72526  72527           PHN  Gene from NCBI      88653\n",
       "72527  72528  LOC143923313  Gene from NCBI      88655\n",
       "72528  72529           FL2  Gene from NCBI      88658\n",
       "\n",
       "[72529 rows x 4 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Used objects from step 2.2\n",
    "stmt = select(\n",
    "    genes.id,\n",
    "    genes.symbol,\n",
    "    genes.hgnc_status,\n",
    "    genes.entity_id\n",
    ")\n",
    "\n",
    "df = bf.query.run_model(stmt)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26692bdf-c42f-45b5-b190-b11831aca93a",
   "metadata": {},
   "source": [
    "#### 3.7. Join two models\n",
    "Joins allow you to combine information from related models using standard SQLAlchemy syntax."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "0bc747ec-c50a-4277-8edc-361dc16c3326",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene symbol</th>\n",
       "      <th>source</th>\n",
       "      <th>alias</th>\n",
       "      <th>type alias</th>\n",
       "      <th>principal alias</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>symbol</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>code</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>code</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENTREZ</td>\n",
       "      <td>1</td>\n",
       "      <td>code</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>UCSC</td>\n",
       "      <td>uc002qsd.5</td>\n",
       "      <td>code</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>synonym</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>symbol</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>HGNC:37133</td>\n",
       "      <td>code</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene symbol   source                   alias type alias  principal alias\n",
       "0        A1BG     HGNC                    A1BG     symbol             True\n",
       "1        A1BG     HGNC                  HGNC:5       code            False\n",
       "2        A1BG  ENSEMBL         ENSG00000121410       code            False\n",
       "3        A1BG   ENTREZ                       1       code            False\n",
       "4        A1BG     UCSC              uc002qsd.5       code            False\n",
       "5        A1BG     HGNC  alpha-1-B glycoprotein    synonym            False\n",
       "6    A1BG-AS1     HGNC                A1BG-AS1     symbol             True\n",
       "7    A1BG-AS1     HGNC              HGNC:37133       code            False"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GeneMaster = bf.query.GeneMaster\n",
    "Alias = bf.query.EntityAlias  # melhor nome\n",
    "\n",
    "stmt = (\n",
    "    bf.query.select(\n",
    "        GeneMaster.symbol.label(\"gene symbol\"),\n",
    "        # Alias.entity_id.label(\"entity_id\"),\n",
    "        Alias.xref_source.label(\"source\"),\n",
    "        Alias.alias_value.label(\"alias\"),\n",
    "        Alias.alias_type.label(\"type alias\"),\n",
    "        Alias.is_primary.label(\"principal alias\"),\n",
    "    )\n",
    "    .select_from(GeneMaster)\n",
    "    .join(Alias, Alias.entity_id == GeneMaster.entity_id)\n",
    "     # .where(Alias.is_primary.is_(True)) # to filter only principal alias\n",
    "    .limit(8)\n",
    ")\n",
    "\n",
    "df = bf.query.run_model(stmt)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "739aa72b-63b1-4bdf-9048-cf46c2d14fb5",
   "metadata": {},
   "source": [
    "#### 3.8. Order results\n",
    "Use order_by() to control sorting of query results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b97e4a61-01d7-4f23-8018-bf933e2e7f73",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>hgnc_status</th>\n",
       "      <th>omic_status_id</th>\n",
       "      <th>entity_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>data_source_id</th>\n",
       "      <th>etl_package_id</th>\n",
       "      <th>locus_group_id</th>\n",
       "      <th>locus_type_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>57864</td>\n",
       "      <td>A-GAMMA3'E</td>\n",
       "      <td>Gene from NCBI</td>\n",
       "      <td>1</td>\n",
       "      <td>57864</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12</td>\n",
       "      <td>A12M1</td>\n",
       "      <td>Entry Withdrawn</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13</td>\n",
       "      <td>A12M2</td>\n",
       "      <td>Entry Withdrawn</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14</td>\n",
       "      <td>A12M3</td>\n",
       "      <td>Entry Withdrawn</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15</td>\n",
       "      <td>A12M4</td>\n",
       "      <td>Entry Withdrawn</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>4</td>\n",
       "      <td>A2M</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>5</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      id      symbol      hgnc_status  omic_status_id  entity_id chromosome  \\\n",
       "0  57864  A-GAMMA3'E   Gene from NCBI               1      57864         11   \n",
       "1     12       A12M1  Entry Withdrawn               1         12       None   \n",
       "2     13       A12M2  Entry Withdrawn               1         13       None   \n",
       "3     14       A12M3  Entry Withdrawn               1         14       None   \n",
       "4     15       A12M4  Entry Withdrawn               1         15       None   \n",
       "5      1        A1BG         Approved               1          1         19   \n",
       "6      2    A1BG-AS1         Approved               1          2         19   \n",
       "7      3        A1CF         Approved               1          3         10   \n",
       "8      4         A2M         Approved               1          4         12   \n",
       "9      5     A2M-AS1         Approved               1          5         12   \n",
       "\n",
       "   data_source_id  etl_package_id  locus_group_id  locus_type_id  \n",
       "0               1               9               9              3  \n",
       "1               2               6               3              3  \n",
       "2               2               6               3              3  \n",
       "3               2               6               3              3  \n",
       "4               2               6               3              3  \n",
       "5               2               6               1              1  \n",
       "6               2               6               2              2  \n",
       "7               2               6               1              1  \n",
       "8               2               6               1              1  \n",
       "9               2               6               2              2  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stmt = (\n",
    "    bf.query.select(bf.query.GeneMaster)\n",
    "    .order_by(bf.query.GeneMaster.symbol)\n",
    "    .limit(10)\n",
    ")\n",
    "\n",
    "df = bf.query.run_model(stmt)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86597ab5-7d97-427a-8950-2e5ff3a8c18e",
   "metadata": {},
   "source": [
    "#### 3.9. Count records\n",
    "Aggregate queries work the same way and return a DataFrame with the computed values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d3c481f8-9ed4-4206-b4c4-0eaeaa1c1f1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>72529</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   value\n",
       "0  72529"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stmt = bf.query.select(\n",
    "    bf.query.func.count(bf.query.GeneMaster.id)\n",
    ")\n",
    "\n",
    "df = bf.query.run_model(stmt)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6e6b3b3-0d47-456f-933b-0126c419e01b",
   "metadata": {},
   "source": [
    "#### 3.10. Use model() as a safe exploratory tool\n",
    "run_model() always returns a DataFrame and catches database errors internally, making it safe for exploratory analysis without breaking the notebook.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "973aac96-9925-4671-aa52-8d13071a22ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>disease_id</th>\n",
       "      <th>label</th>\n",
       "      <th>description</th>\n",
       "      <th>omic_status_id</th>\n",
       "      <th>entity_id</th>\n",
       "      <th>data_source_id</th>\n",
       "      <th>etl_package_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>MONDO:0000002</td>\n",
       "      <td>obsolete 46,XX sex reversal</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>114901</td>\n",
       "      <td>37</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>MONDO:0000003</td>\n",
       "      <td>obsolete 17-hydroxysteroid dehydrogenase defic...</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>114904</td>\n",
       "      <td>37</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>MONDO:0000004</td>\n",
       "      <td>adrenocortical insufficiency</td>\n",
       "      <td>An endocrine or hormonal disorder that occurs ...</td>\n",
       "      <td>1</td>\n",
       "      <td>114907</td>\n",
       "      <td>37</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>MONDO:0000005</td>\n",
       "      <td>alopecia, isolated</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>114912</td>\n",
       "      <td>37</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>MONDO:0000006</td>\n",
       "      <td>obsolete alopecia-mental retardation syndrome</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>114915</td>\n",
       "      <td>37</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id     disease_id                                              label  \\\n",
       "0   1  MONDO:0000002                        obsolete 46,XX sex reversal   \n",
       "1   2  MONDO:0000003  obsolete 17-hydroxysteroid dehydrogenase defic...   \n",
       "2   3  MONDO:0000004                       adrenocortical insufficiency   \n",
       "3   4  MONDO:0000005                                 alopecia, isolated   \n",
       "4   5  MONDO:0000006      obsolete alopecia-mental retardation syndrome   \n",
       "\n",
       "                                         description  omic_status_id  \\\n",
       "0                                                                  2   \n",
       "1                                                                  2   \n",
       "2  An endocrine or hormonal disorder that occurs ...               1   \n",
       "3                                                                  1   \n",
       "4                                                                  2   \n",
       "\n",
       "   entity_id  data_source_id  etl_package_id  \n",
       "0     114901              37              24  \n",
       "1     114904              37              24  \n",
       "2     114907              37              24  \n",
       "3     114912              37              24  \n",
       "4     114915              37              24  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = bf.query.run_model(\n",
    "    bf.query.select(bf.query.DiseaseMaster).limit(5)\n",
    ")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dca7b24a-b8b4-4050-b81f-be16e610bb2a",
   "metadata": {},
   "source": [
    "-------------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97d42e64-5d05-42d8-bd3e-6817b7fdeb42",
   "metadata": {},
   "source": [
    "### 4. Select data using SQL code\n",
    "\n",
    "The run_sql() method executes raw SQL statements directly against the Biofilter3R database and always returns the results as a pandas DataFrame.\n",
    "It is intended for advanced users, debugging, validation, or situations where writing SQL is faster or more expressive than building a SQLAlchemy statement.\n",
    "\n",
    "Unlike run_model(), run_sql() bypasses the ORM layer entirely.\n",
    "\n",
    "#### When to use run_sql()\n",
    "\n",
    "Use run_sql() when you need to:\n",
    "\n",
    "* Quickly inspect a table using plain SQL\n",
    "* Debug or validate ORM-generated queries\n",
    "* Run database-specific SQL (CTEs, window functions, vendor-specific syntax)\n",
    "* Explore indexes, counts, or performance-related questions\n",
    "* Copy-paste queries from database consoles (psql, pgAdmin, etc.)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd3c147b-fc1e-41bc-8a7b-fccb6c2ed4de",
   "metadata": {},
   "source": [
    "#### 4.1 Preview an entire table\n",
    "This is the fastest way to inspect raw table contents without using ORM models."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "97ebac48-3c24-4a5f-b370-61028d9cf634",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>hgnc_status</th>\n",
       "      <th>omic_status_id</th>\n",
       "      <th>entity_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>data_source_id</th>\n",
       "      <th>etl_package_id</th>\n",
       "      <th>locus_group_id</th>\n",
       "      <th>locus_type_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>A2M</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>A2ML1-AS1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>A2ML1-AS2</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>22</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id     symbol hgnc_status  omic_status_id  entity_id chromosome  \\\n",
       "0   1       A1BG    Approved               1          1         19   \n",
       "1   2   A1BG-AS1    Approved               1          2         19   \n",
       "2   3       A1CF    Approved               1          3         10   \n",
       "3   4        A2M    Approved               1          4         12   \n",
       "4   5    A2M-AS1    Approved               1          5         12   \n",
       "5   6      A2ML1    Approved               1          6         12   \n",
       "6   7  A2ML1-AS1    Approved               1          7         12   \n",
       "7   8  A2ML1-AS2    Approved               1          8         12   \n",
       "8   9    A3GALT2    Approved               1          9          1   \n",
       "9  10     A4GALT    Approved               1         10         22   \n",
       "\n",
       "   data_source_id  etl_package_id  locus_group_id  locus_type_id  \n",
       "0               2               6               1              1  \n",
       "1               2               6               2              2  \n",
       "2               2               6               1              1  \n",
       "3               2               6               1              1  \n",
       "4               2               6               2              2  \n",
       "5               2               6               1              1  \n",
       "6               2               6               2              2  \n",
       "7               2               6               2              2  \n",
       "8               2               6               1              1  \n",
       "9               2               6               1              1  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = bf.query.run_sql(\n",
    "    \"SELECT * FROM gene_masters LIMIT 10\"\n",
    ")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e058a0f8-35d9-49cd-a4a3-0f26a7b300c5",
   "metadata": {},
   "source": [
    "#### 4.2 Adding complexity with a more advanced query.\n",
    "In this example, we retrieve all relationships between Genes and Diseases and enrich the results by including the genomic locations of the Genes mapped to genome assembly GRCh38."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "cb5d11b8-0135-4481-8914-4859bedfe5f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>build</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>end_pos</th>\n",
       "      <th>relationship_type</th>\n",
       "      <th>disease_name</th>\n",
       "      <th>source_system</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RGS9</td>\n",
       "      <td>38</td>\n",
       "      <td>17</td>\n",
       "      <td>65100812</td>\n",
       "      <td>65227703</td>\n",
       "      <td>part_of</td>\n",
       "      <td>bradyopsia</td>\n",
       "      <td>CLINGEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RHO</td>\n",
       "      <td>38</td>\n",
       "      <td>3</td>\n",
       "      <td>129528639</td>\n",
       "      <td>129535344</td>\n",
       "      <td>part_of</td>\n",
       "      <td>inherited retinal dystrophy</td>\n",
       "      <td>CLINGEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GRK1</td>\n",
       "      <td>38</td>\n",
       "      <td>13</td>\n",
       "      <td>113667219</td>\n",
       "      <td>113737736</td>\n",
       "      <td>part_of</td>\n",
       "      <td>Oguchi disease</td>\n",
       "      <td>CLINGEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RIPK1</td>\n",
       "      <td>38</td>\n",
       "      <td>6</td>\n",
       "      <td>3063824</td>\n",
       "      <td>3115187</td>\n",
       "      <td>part_of</td>\n",
       "      <td>immunodeficiency 57</td>\n",
       "      <td>CLINGEN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RIT1</td>\n",
       "      <td>38</td>\n",
       "      <td>1</td>\n",
       "      <td>155897808</td>\n",
       "      <td>155911408</td>\n",
       "      <td>part_of</td>\n",
       "      <td>Noonan syndrome</td>\n",
       "      <td>CLINGEN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_symbol  build  chromosome  start_pos    end_pos relationship_type  \\\n",
       "0        RGS9     38          17   65100812   65227703           part_of   \n",
       "1         RHO     38           3  129528639  129535344           part_of   \n",
       "2        GRK1     38          13  113667219  113737736           part_of   \n",
       "3       RIPK1     38           6    3063824    3115187           part_of   \n",
       "4        RIT1     38           1  155897808  155911408           part_of   \n",
       "\n",
       "                  disease_name source_system  \n",
       "0                   bradyopsia       CLINGEN  \n",
       "1  inherited retinal dystrophy       CLINGEN  \n",
       "2               Oguchi disease       CLINGEN  \n",
       "3          immunodeficiency 57       CLINGEN  \n",
       "4              Noonan syndrome       CLINGEN  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sql = \"\"\"\n",
    "WITH groups AS (\n",
    "  SELECT\n",
    "    MAX(CASE WHEN name = 'Genes' THEN id END)     AS gene_group_id,\n",
    "    MAX(CASE WHEN name = 'Diseases' THEN id END)  AS disease_group_id\n",
    "  FROM entity_groups\n",
    "),\n",
    "gd_rels AS (\n",
    "  SELECT\n",
    "    er.*,\n",
    "\n",
    "    CASE\n",
    "      WHEN er.entity_1_group_id = g.gene_group_id THEN er.entity_1_id\n",
    "      ELSE er.entity_2_id\n",
    "    END AS gene_entity_id,\n",
    "\n",
    "    CASE\n",
    "      WHEN er.entity_1_group_id = g.disease_group_id THEN er.entity_1_id\n",
    "      ELSE er.entity_2_id\n",
    "    END AS disease_entity_id\n",
    "\n",
    "  FROM entity_relationships er\n",
    "  CROSS JOIN groups g\n",
    "  WHERE\n",
    "    (er.entity_1_group_id = g.gene_group_id AND er.entity_2_group_id = g.disease_group_id)\n",
    "    OR\n",
    "    (er.entity_1_group_id = g.disease_group_id AND er.entity_2_group_id = g.gene_group_id)\n",
    ")\n",
    "\n",
    "SELECT\n",
    "  -- r.id AS relationship_id,\n",
    "\n",
    "  -- Gene\n",
    "  -- r.gene_entity_id,\n",
    "  -- 'Gene' AS gene_group,\n",
    "  ga.alias_value AS gene_symbol,\n",
    "\n",
    "  -- Gene location (GRCh38)\n",
    "  -- el.assembly_id,\n",
    "  el.build,\n",
    "  el.chromosome,\n",
    "  el.start_pos,\n",
    "  el.end_pos,\n",
    "  -- el.strand,\n",
    "  -- el.region_label,\n",
    "\n",
    "  -- Relationship\n",
    "  rt.code AS relationship_type,\n",
    "  \n",
    "  -- Disease\n",
    "  -- r.disease_entity_id,\n",
    "  -- 'Disease' AS disease_group,\n",
    "  -- dm.disease_id AS disease_id,\n",
    "  COALESCE(dm.label, da.alias_value) AS disease_name,\n",
    "\n",
    "  -- Provenance\n",
    "  -- ds.name AS data_source,\n",
    "  ss.name AS source_system\n",
    "\n",
    "FROM gd_rels r\n",
    "JOIN entity_relationship_types rt\n",
    "  ON rt.id = r.relationship_type_id\n",
    "\n",
    "-- Gene primary alias\n",
    "LEFT JOIN entity_aliases ga\n",
    "  ON ga.entity_id = r.gene_entity_id\n",
    " AND (ga.is_primary = TRUE OR ga.alias_type = 'preferred')\n",
    "\n",
    "-- Disease alias (usually MONDO code)\n",
    "LEFT JOIN entity_aliases da\n",
    "  ON da.entity_id = r.disease_entity_id\n",
    " AND (da.is_primary = TRUE OR da.alias_type = 'preferred')\n",
    "\n",
    "-- Disease master (human-readable label)\n",
    "LEFT JOIN disease_masters dm\n",
    "  ON dm.entity_id = r.disease_entity_id\n",
    "\n",
    "-- Gene location\n",
    "LEFT JOIN entity_locations el\n",
    "  ON el.entity_id = r.gene_entity_id\n",
    " AND el.build = 38\n",
    "\n",
    "-- Provenance joins\n",
    "LEFT JOIN etl_data_sources ds\n",
    "  ON ds.id = r.data_source_id\n",
    "LEFT JOIN etl_source_systems ss\n",
    "  ON ss.id = ds.source_system_id\n",
    "\n",
    "ORDER BY r.id\n",
    "LIMIT 1000\n",
    "\"\"\"\n",
    "df = bf.query.run_sql(sql)\n",
    "df.head()\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f6d7a3d-1190-419f-8240-0d4139bf766b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
