{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "567f7631-4c2a-4873-8254-6a0c861931c1",
   "metadata": {},
   "source": [
    "<h1> üìò Biofilter3R ‚Äî Reports: Gene ‚Üí SNP Annotation </h1>\n",
    "\n",
    "This tutorial shows how to use the Reports API to generate standardized annotation outputs from Genes ‚Üí SNPs using Biofilter.\n",
    "The main report in this notebook resolves SNPs by genomic overlap between EntityLocations (genes, build 38) and VariantSNP (rsID positions, build 37/38).\n",
    "\n",
    "### Methods\n",
    "\n",
    "* report.list() ‚Äî List all available reports.\n",
    "* report.explain(\"<report>\") ‚Äî Print a detailed explanation of a report (inputs, outputs, rules).\n",
    "* report.available_columns(\"<report>\") ‚Äî Show all optional output columns supported by the report.\n",
    "* report.run(\"<report>\", **params) ‚Äî Execute a report and return a pandas DataFrame.\n",
    "* report.run_example(\"<report>\") ‚Äî Run a report using its built-in example_input().\n",
    "\n",
    "### Available Reports\n",
    "\n",
    "* gene_to_snp ‚Äî Given a list of genes (symbols/IDs/aliases), returns SNPs overlapping each gene region (Build 38 gene coordinates + SNP positions in Build 38 and/or Build 37)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88fc0a71-bd81-4a7b-9a50-8e457debc956",
   "metadata": {},
   "source": [
    "--------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d07f422f-027d-41d0-8adb-df318a662047",
   "metadata": {},
   "source": [
    "### Start Biofilter3R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c641f608-ecc9-48d1-a9cb-8e4208be9398",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[INFO] ‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê\n",
      "[INFO] üöÄ Initializing Biofilter3R\n",
      "[INFO]    ‚Ä¢ Version: 3.2.0\n",
      "[INFO]    ‚Ä¢ Debug mode: False\n",
      "[INFO]    ‚Ä¢ Config: /home/bioadmin/biofilter/.biofilter.toml\n",
      "[INFO] ‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê\n",
      "[INFO] üîå Database connection established\n",
      "[INFO]    ‚Ä¢ Engine: postgresql+psycopg2\n",
      "[INFO]    ‚Ä¢ Host:   localhost\n",
      "[INFO]    ‚Ä¢ DB:     biofilter\n",
      "[INFO]    ‚Ä¢ Time:   1.3 ms\n",
      "[INFO] ‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê\n"
     ]
    }
   ],
   "source": [
    "from biofilter import Biofilter\n",
    "bf = Biofilter()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b55ab3ee-cd92-4bb2-ab9a-07ab37956681",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Œî: 290ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Input Gene</th>\n",
       "      <th>Matched Name</th>\n",
       "      <th>Alias Type</th>\n",
       "      <th>Alias Source</th>\n",
       "      <th>HGNC Symbol</th>\n",
       "      <th>Gene Entity ID</th>\n",
       "      <th>Gene Build</th>\n",
       "      <th>Gene Chr (23:X/24:Y)</th>\n",
       "      <th>Gene Start (Build 38)</th>\n",
       "      <th>Gene End (Build 38)</th>\n",
       "      <th>...</th>\n",
       "      <th>Variant Type</th>\n",
       "      <th>Variant ID</th>\n",
       "      <th>SNP Chr (23:X/24:Y)</th>\n",
       "      <th>SNP Pos (Build 38)</th>\n",
       "      <th>SNP Pos (Build 37)</th>\n",
       "      <th>Ref Allele</th>\n",
       "      <th>Alt Allele</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>SourceSystem</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>code</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>58345178</td>\n",
       "      <td>58353492</td>\n",
       "      <td>...</td>\n",
       "      <td>rs</td>\n",
       "      <td>1451</td>\n",
       "      <td>19</td>\n",
       "      <td>58354491</td>\n",
       "      <td>58865857</td>\n",
       "      <td>G</td>\n",
       "      <td>A/C/T</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>code</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>58345178</td>\n",
       "      <td>58353492</td>\n",
       "      <td>...</td>\n",
       "      <td>rs</td>\n",
       "      <td>893180</td>\n",
       "      <td>19</td>\n",
       "      <td>58353725</td>\n",
       "      <td>58865091</td>\n",
       "      <td>A</td>\n",
       "      <td>C/G/T</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>code</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>58345178</td>\n",
       "      <td>58353492</td>\n",
       "      <td>...</td>\n",
       "      <td>rs</td>\n",
       "      <td>893181</td>\n",
       "      <td>19</td>\n",
       "      <td>58353716</td>\n",
       "      <td>58865082</td>\n",
       "      <td>A</td>\n",
       "      <td>C/G/T</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>code</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>58345178</td>\n",
       "      <td>58353492</td>\n",
       "      <td>...</td>\n",
       "      <td>rs</td>\n",
       "      <td>893182</td>\n",
       "      <td>19</td>\n",
       "      <td>58353702</td>\n",
       "      <td>58865068</td>\n",
       "      <td>T</td>\n",
       "      <td>A/C</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>code</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>58345178</td>\n",
       "      <td>58353492</td>\n",
       "      <td>...</td>\n",
       "      <td>rs</td>\n",
       "      <td>893183</td>\n",
       "      <td>19</td>\n",
       "      <td>58353688</td>\n",
       "      <td>58865054</td>\n",
       "      <td>T</td>\n",
       "      <td>A/C/G</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows √ó 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Input Gene Matched Name Alias Type Alias Source HGNC Symbol  Gene Entity ID  \\\n",
       "0     HGNC:5       HGNC:5       code         HGNC        A1BG               1   \n",
       "1     HGNC:5       HGNC:5       code         HGNC        A1BG               1   \n",
       "2     HGNC:5       HGNC:5       code         HGNC        A1BG               1   \n",
       "3     HGNC:5       HGNC:5       code         HGNC        A1BG               1   \n",
       "4     HGNC:5       HGNC:5       code         HGNC        A1BG               1   \n",
       "\n",
       "  Gene Build Gene Chr (23:X/24:Y) Gene Start (Build 38) Gene End (Build 38)  \\\n",
       "0         38                   19              58345178            58353492   \n",
       "1         38                   19              58345178            58353492   \n",
       "2         38                   19              58345178            58353492   \n",
       "3         38                   19              58345178            58353492   \n",
       "4         38                   19              58345178            58353492   \n",
       "\n",
       "   ... Variant Type Variant ID SNP Chr (23:X/24:Y) SNP Pos (Build 38)  \\\n",
       "0  ...           rs       1451                  19           58354491   \n",
       "1  ...           rs     893180                  19           58353725   \n",
       "2  ...           rs     893181                  19           58353716   \n",
       "3  ...           rs     893182                  19           58353702   \n",
       "4  ...           rs     893183                  19           58353688   \n",
       "\n",
       "  SNP Pos (Build 37) Ref Allele Alt Allele   DataSource SourceSystem Note  \n",
       "0           58865857          G      A/C/T  dbsnp_chr19         NCBI  NaN  \n",
       "1           58865091          A      C/G/T  dbsnp_chr19         NCBI  NaN  \n",
       "2           58865082          A      C/G/T  dbsnp_chr19         NCBI  NaN  \n",
       "3           58865068          T        A/C  dbsnp_chr19         NCBI  NaN  \n",
       "4           58865054          T      A/C/G  dbsnp_chr19         NCBI  NaN  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Before\n",
    "import time\n",
    "start_time = time.time()\n",
    "\n",
    "df = bf.report.run_example(\"gene_to_snp\")\n",
    "end_time = time.time() - start_time\n",
    "print(f\"Œî: {end_time:.0f}ms\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7b99723e-b506-44a6-81d7-b33dd296df40",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Œî: 0s\n",
      "7961\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Input Gene</th>\n",
       "      <th>Matched Name</th>\n",
       "      <th>Alias Type</th>\n",
       "      <th>Alias Source</th>\n",
       "      <th>HGNC Symbol</th>\n",
       "      <th>Gene Entity ID</th>\n",
       "      <th>Gene Build</th>\n",
       "      <th>Gene Chr (23:X/24:Y)</th>\n",
       "      <th>Gene Start (Build 38)</th>\n",
       "      <th>Gene End (Build 38)</th>\n",
       "      <th>...</th>\n",
       "      <th>Variant Type</th>\n",
       "      <th>Variant ID</th>\n",
       "      <th>SNP Chr (23:X/24:Y)</th>\n",
       "      <th>SNP Pos (Build 38)</th>\n",
       "      <th>SNP Pos (Build 37)</th>\n",
       "      <th>Ref Allele</th>\n",
       "      <th>Alt Allele</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>SourceSystem</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>code</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>58345178</td>\n",
       "      <td>58353492</td>\n",
       "      <td>...</td>\n",
       "      <td>rs</td>\n",
       "      <td>575574507</td>\n",
       "      <td>19</td>\n",
       "      <td>58344180</td>\n",
       "      <td>58855546</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>code</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>58345178</td>\n",
       "      <td>58353492</td>\n",
       "      <td>...</td>\n",
       "      <td>rs</td>\n",
       "      <td>375555639</td>\n",
       "      <td>19</td>\n",
       "      <td>58344182</td>\n",
       "      <td>58855548</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>code</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>58345178</td>\n",
       "      <td>58353492</td>\n",
       "      <td>...</td>\n",
       "      <td>rs</td>\n",
       "      <td>1418884653</td>\n",
       "      <td>19</td>\n",
       "      <td>58344187</td>\n",
       "      <td>58855553</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>code</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>58345178</td>\n",
       "      <td>58353492</td>\n",
       "      <td>...</td>\n",
       "      <td>rs</td>\n",
       "      <td>138605506</td>\n",
       "      <td>19</td>\n",
       "      <td>58344188</td>\n",
       "      <td>58855554</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>code</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>58345178</td>\n",
       "      <td>58353492</td>\n",
       "      <td>...</td>\n",
       "      <td>rs</td>\n",
       "      <td>2051900169</td>\n",
       "      <td>19</td>\n",
       "      <td>58344193</td>\n",
       "      <td>58855559</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows √ó 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Input Gene Matched Name Alias Type Alias Source HGNC Symbol  Gene Entity ID  \\\n",
       "0     HGNC:5       HGNC:5       code         HGNC        A1BG               1   \n",
       "1     HGNC:5       HGNC:5       code         HGNC        A1BG               1   \n",
       "2     HGNC:5       HGNC:5       code         HGNC        A1BG               1   \n",
       "3     HGNC:5       HGNC:5       code         HGNC        A1BG               1   \n",
       "4     HGNC:5       HGNC:5       code         HGNC        A1BG               1   \n",
       "\n",
       "  Gene Build Gene Chr (23:X/24:Y) Gene Start (Build 38) Gene End (Build 38)  \\\n",
       "0         38                   19              58345178            58353492   \n",
       "1         38                   19              58345178            58353492   \n",
       "2         38                   19              58345178            58353492   \n",
       "3         38                   19              58345178            58353492   \n",
       "4         38                   19              58345178            58353492   \n",
       "\n",
       "   ... Variant Type  Variant ID SNP Chr (23:X/24:Y) SNP Pos (Build 38)  \\\n",
       "0  ...           rs   575574507                  19           58344180   \n",
       "1  ...           rs   375555639                  19           58344182   \n",
       "2  ...           rs  1418884653                  19           58344187   \n",
       "3  ...           rs   138605506                  19           58344188   \n",
       "4  ...           rs  2051900169                  19           58344193   \n",
       "\n",
       "  SNP Pos (Build 37) Ref Allele Alt Allele   DataSource SourceSystem Note  \n",
       "0           58855546          C          T  dbsnp_chr19         NCBI  NaN  \n",
       "1           58855548          T          C  dbsnp_chr19         NCBI  NaN  \n",
       "2           58855553          G          A  dbsnp_chr19         NCBI  NaN  \n",
       "3           58855554          C          G  dbsnp_chr19         NCBI  NaN  \n",
       "4           58855559          A          T  dbsnp_chr19         NCBI  NaN  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# After\n",
    "import time\n",
    "start_time = time.time()\n",
    "\n",
    "df = bf.report.run_example(\"gene_to_snp\")\n",
    "end_time = time.time() - start_time\n",
    "print(f\"Œî: {end_time:.0f}s\")\n",
    "print(len(df))\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b3690ad-7734-4fd6-83c3-4b73ff142a52",
   "metadata": {},
   "source": [
    "------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "feb45e5a-66c5-4baa-980b-1552c156b103",
   "metadata": {},
   "source": [
    "### 1. Inspect the report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "623e012c-b1e2-404a-9432-34568adf7a3f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üß¨ GENE ‚Üí SNP Report (v3.2.0)\n",
      "============================\n",
      "\n",
      "This report takes gene identifiers (symbols, HGNC IDs, Entrez IDs, Ensembl IDs, synonyms)\n",
      "and returns SNPs overlapping each gene region.\n",
      "\n",
      "Rules:\n",
      "- Gene regions come from EntityLocation and are currently available only for build 38.\n",
      "- SNP lookup uses VariantSNP and overlaps are computed using build 38 coordinates (position_38).\n",
      "- VariantSNP contains both position_38 and position_37; the report can display both.\n",
      "\n",
      "Parameters:\n",
      "- input_data: list[str] or path to .txt file (required)\n",
      "- window_bp: int >= 0 (default: 1000). Extends the gene region +/- window_bp (build 38).\n",
      "- assembly: \"37\" | \"38\" | None (default None).\n",
      "    NOTE: assembly is OUTPUT-ONLY:\n",
      "      - None: show both 'SNP Pos (Build 38)' and 'SNP Pos (Build 37)'\n",
      "      - \"38\": show only 'SNP Pos (Build 38)'\n",
      "      - \"37\": show only 'SNP Pos (Build 37)'\n",
      "- output_columns: optional list[str] of DISPLAY NAMES (see available_columns()).\n",
      "\n",
      "Usage:\n",
      "    df = bf.report.run(\n",
      "        \"gene_to_snp\",\n",
      "        input_data=[\"TP53\", \"HGNC:11998\"],\n",
      "        window_bp=5000,\n",
      "        assembly=None,\n",
      "        output_columns=[\n",
      "            \"Input Gene\",\n",
      "            \"HGNC Symbol\",\n",
      "            \"Gene Chr (23:X/24:Y)\",\n",
      "            \"Gene Start (Build 38)\",\n",
      "            \"Gene End (Build 38)\",\n",
      "            \"Variant Type\",\n",
      "            \"Variant ID\",\n",
      "            \"SNP Pos (Build 38)\",\n",
      "            \"SNP Pos (Build 37)\",\n",
      "            \"Note\",\n",
      "        ],\n",
      "    )\n",
      "\n"
     ]
    }
   ],
   "source": [
    "bf.report.explain(\"gene_to_snp\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f90c1317-e28f-4d6c-adc9-6e47dcf9de12",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Input Gene', 'Matched Name', 'Alias Type', 'Alias Source', 'HGNC Symbol', 'Gene Entity ID', 'Gene Build', 'Gene Chr (23:X/24:Y)', 'Gene Start (Build 38)', 'Gene End (Build 38)', 'Gene Strand', 'Region Label', 'Variant Type', 'Variant ID', 'SNP Chr (23:X/24:Y)', 'SNP Pos (Build 38)', 'SNP Pos (Build 37)', 'Ref Allele', 'Alt Allele', 'DataSource', 'SourceSystem', 'Note']\n"
     ]
    }
   ],
   "source": [
    "bf.report.available_columns(\"gene_to_snp\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2eff93e6-d143-483c-b1e6-91b65d3062be",
   "metadata": {},
   "source": [
    "---------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bbe4e7b-14fa-4722-acca-650d82b66707",
   "metadata": {},
   "source": [
    "### 2. Run a simple example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "01b81fb8-3ab1-47e1-9cbd-cd267d79e0be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Input Gene</th>\n",
       "      <th>Matched Name</th>\n",
       "      <th>Alias Type</th>\n",
       "      <th>Alias Source</th>\n",
       "      <th>HGNC Symbol</th>\n",
       "      <th>Gene Entity ID</th>\n",
       "      <th>Gene Build</th>\n",
       "      <th>Gene Chr (23:X/24:Y)</th>\n",
       "      <th>Gene Start (Build 38)</th>\n",
       "      <th>Gene End (Build 38)</th>\n",
       "      <th>...</th>\n",
       "      <th>Variant Type</th>\n",
       "      <th>Variant ID</th>\n",
       "      <th>SNP Chr (23:X/24:Y)</th>\n",
       "      <th>SNP Pos (Build 38)</th>\n",
       "      <th>SNP Pos (Build 37)</th>\n",
       "      <th>Ref Allele</th>\n",
       "      <th>Alt Allele</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>SourceSystem</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>code</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>58345178</td>\n",
       "      <td>58353492</td>\n",
       "      <td>...</td>\n",
       "      <td>rs</td>\n",
       "      <td>1451</td>\n",
       "      <td>19</td>\n",
       "      <td>58354491</td>\n",
       "      <td>58865857</td>\n",
       "      <td>G</td>\n",
       "      <td>A/C/T</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>code</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>58345178</td>\n",
       "      <td>58353492</td>\n",
       "      <td>...</td>\n",
       "      <td>rs</td>\n",
       "      <td>893180</td>\n",
       "      <td>19</td>\n",
       "      <td>58353725</td>\n",
       "      <td>58865091</td>\n",
       "      <td>A</td>\n",
       "      <td>C/G/T</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>code</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>58345178</td>\n",
       "      <td>58353492</td>\n",
       "      <td>...</td>\n",
       "      <td>rs</td>\n",
       "      <td>893181</td>\n",
       "      <td>19</td>\n",
       "      <td>58353716</td>\n",
       "      <td>58865082</td>\n",
       "      <td>A</td>\n",
       "      <td>C/G/T</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>code</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>58345178</td>\n",
       "      <td>58353492</td>\n",
       "      <td>...</td>\n",
       "      <td>rs</td>\n",
       "      <td>893182</td>\n",
       "      <td>19</td>\n",
       "      <td>58353702</td>\n",
       "      <td>58865068</td>\n",
       "      <td>T</td>\n",
       "      <td>A/C</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>code</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>58345178</td>\n",
       "      <td>58353492</td>\n",
       "      <td>...</td>\n",
       "      <td>rs</td>\n",
       "      <td>893183</td>\n",
       "      <td>19</td>\n",
       "      <td>58353688</td>\n",
       "      <td>58865054</td>\n",
       "      <td>T</td>\n",
       "      <td>A/C/G</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows √ó 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Input Gene Matched Name Alias Type Alias Source HGNC Symbol  Gene Entity ID  \\\n",
       "0     HGNC:5       HGNC:5       code         HGNC        A1BG               1   \n",
       "1     HGNC:5       HGNC:5       code         HGNC        A1BG               1   \n",
       "2     HGNC:5       HGNC:5       code         HGNC        A1BG               1   \n",
       "3     HGNC:5       HGNC:5       code         HGNC        A1BG               1   \n",
       "4     HGNC:5       HGNC:5       code         HGNC        A1BG               1   \n",
       "\n",
       "  Gene Build Gene Chr (23:X/24:Y) Gene Start (Build 38) Gene End (Build 38)  \\\n",
       "0         38                   19              58345178            58353492   \n",
       "1         38                   19              58345178            58353492   \n",
       "2         38                   19              58345178            58353492   \n",
       "3         38                   19              58345178            58353492   \n",
       "4         38                   19              58345178            58353492   \n",
       "\n",
       "   ... Variant Type Variant ID SNP Chr (23:X/24:Y) SNP Pos (Build 38)  \\\n",
       "0  ...           rs       1451                  19           58354491   \n",
       "1  ...           rs     893180                  19           58353725   \n",
       "2  ...           rs     893181                  19           58353716   \n",
       "3  ...           rs     893182                  19           58353702   \n",
       "4  ...           rs     893183                  19           58353688   \n",
       "\n",
       "  SNP Pos (Build 37) Ref Allele Alt Allele   DataSource SourceSystem Note  \n",
       "0           58865857          G      A/C/T  dbsnp_chr19         NCBI  NaN  \n",
       "1           58865091          A      C/G/T  dbsnp_chr19         NCBI  NaN  \n",
       "2           58865082          A      C/G/T  dbsnp_chr19         NCBI  NaN  \n",
       "3           58865068          T        A/C  dbsnp_chr19         NCBI  NaN  \n",
       "4           58865054          T      A/C/G  dbsnp_chr19         NCBI  NaN  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = bf.report.run_example(\"gene_to_snp\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "602c9d7d-5ec7-49ab-9f8c-a1ee2d39871c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TXLNGY', 'HGNC:18473', '246126', 'ENSG00000131002', 'HGNC:5']\n"
     ]
    }
   ],
   "source": [
    "bf.report.example_input(\"gene_to_snp\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84df5465-ee70-44d2-8342-8c2a55771458",
   "metadata": {},
   "source": [
    "--------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ee55450-bf43-4ddb-b4cb-3544dcd937bc",
   "metadata": {},
   "source": [
    "### 3. Run with custom inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b3bbecc5-7934-4b0d-b42f-fd49658422ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Input Gene</th>\n",
       "      <th>Matched Name</th>\n",
       "      <th>Alias Type</th>\n",
       "      <th>Alias Source</th>\n",
       "      <th>HGNC Symbol</th>\n",
       "      <th>Gene Entity ID</th>\n",
       "      <th>Gene Build</th>\n",
       "      <th>Gene Chr (23:X/24:Y)</th>\n",
       "      <th>Gene Start (Build 38)</th>\n",
       "      <th>Gene End (Build 38)</th>\n",
       "      <th>...</th>\n",
       "      <th>Region Label</th>\n",
       "      <th>Variant Type</th>\n",
       "      <th>Variant ID</th>\n",
       "      <th>SNP Chr (23:X/24:Y)</th>\n",
       "      <th>SNP Pos (Build 38)</th>\n",
       "      <th>Ref Allele</th>\n",
       "      <th>Alt Allele</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>SourceSystem</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Apoe</td>\n",
       "      <td>APOE</td>\n",
       "      <td>symbol</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>APOE</td>\n",
       "      <td>11434</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>44903787</td>\n",
       "      <td>44909396</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>rs</td>\n",
       "      <td>1467532231</td>\n",
       "      <td>19</td>\n",
       "      <td>44902792</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Apoe</td>\n",
       "      <td>APOE</td>\n",
       "      <td>symbol</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>APOE</td>\n",
       "      <td>11434</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>44903787</td>\n",
       "      <td>44909396</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>rs</td>\n",
       "      <td>2513533110</td>\n",
       "      <td>19</td>\n",
       "      <td>44902793</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Apoe</td>\n",
       "      <td>APOE</td>\n",
       "      <td>symbol</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>APOE</td>\n",
       "      <td>11434</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>44903787</td>\n",
       "      <td>44909396</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>rs</td>\n",
       "      <td>2513533115</td>\n",
       "      <td>19</td>\n",
       "      <td>44902794</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Apoe</td>\n",
       "      <td>APOE</td>\n",
       "      <td>symbol</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>APOE</td>\n",
       "      <td>11434</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>44903787</td>\n",
       "      <td>44909396</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>rs</td>\n",
       "      <td>955657742</td>\n",
       "      <td>19</td>\n",
       "      <td>44902795</td>\n",
       "      <td>C</td>\n",
       "      <td>A/G/T</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Apoe</td>\n",
       "      <td>APOE</td>\n",
       "      <td>symbol</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>APOE</td>\n",
       "      <td>11434</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>44903787</td>\n",
       "      <td>44909396</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>rs</td>\n",
       "      <td>2513533119</td>\n",
       "      <td>19</td>\n",
       "      <td>44902797</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows √ó 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Input Gene Matched Name Alias Type Alias Source HGNC Symbol  Gene Entity ID  \\\n",
       "0       Apoe         APOE     symbol         HGNC        APOE           11434   \n",
       "1       Apoe         APOE     symbol         HGNC        APOE           11434   \n",
       "2       Apoe         APOE     symbol         HGNC        APOE           11434   \n",
       "3       Apoe         APOE     symbol         HGNC        APOE           11434   \n",
       "4       Apoe         APOE     symbol         HGNC        APOE           11434   \n",
       "\n",
       "  Gene Build Gene Chr (23:X/24:Y) Gene Start (Build 38) Gene End (Build 38)  \\\n",
       "0         38                   19              44903787            44909396   \n",
       "1         38                   19              44903787            44909396   \n",
       "2         38                   19              44903787            44909396   \n",
       "3         38                   19              44903787            44909396   \n",
       "4         38                   19              44903787            44909396   \n",
       "\n",
       "   ... Region Label Variant Type  Variant ID SNP Chr (23:X/24:Y)  \\\n",
       "0  ...         None           rs  1467532231                  19   \n",
       "1  ...         None           rs  2513533110                  19   \n",
       "2  ...         None           rs  2513533115                  19   \n",
       "3  ...         None           rs   955657742                  19   \n",
       "4  ...         None           rs  2513533119                  19   \n",
       "\n",
       "  SNP Pos (Build 38) Ref Allele Alt Allele   DataSource SourceSystem  Note  \n",
       "0           44902792          A          G  dbsnp_chr19         NCBI  <NA>  \n",
       "1           44902793          A          G  dbsnp_chr19         NCBI  <NA>  \n",
       "2           44902794          C          T  dbsnp_chr19         NCBI  <NA>  \n",
       "3           44902795          C      A/G/T  dbsnp_chr19         NCBI  <NA>  \n",
       "4           44902797          A          G  dbsnp_chr19         NCBI  <NA>  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Return Variants in Build 38\n",
    "df = bf.report.run(\n",
    "    \"gene_to_snp\",\n",
    "    input_data = ['Tp53', 'Apoe'],\n",
    "    assembly = 38\n",
    ")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4c736009-9167-4e98-b0b6-c42e3b365fad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Input Gene</th>\n",
       "      <th>Matched Name</th>\n",
       "      <th>Alias Type</th>\n",
       "      <th>Alias Source</th>\n",
       "      <th>HGNC Symbol</th>\n",
       "      <th>Gene Entity ID</th>\n",
       "      <th>Gene Build</th>\n",
       "      <th>Gene Chr (23:X/24:Y)</th>\n",
       "      <th>Gene Start (Build 38)</th>\n",
       "      <th>Gene End (Build 38)</th>\n",
       "      <th>...</th>\n",
       "      <th>Region Label</th>\n",
       "      <th>Variant Type</th>\n",
       "      <th>Variant ID</th>\n",
       "      <th>SNP Chr (23:X/24:Y)</th>\n",
       "      <th>SNP Pos (Build 37)</th>\n",
       "      <th>Ref Allele</th>\n",
       "      <th>Alt Allele</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>SourceSystem</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Apoe</td>\n",
       "      <td>APOE</td>\n",
       "      <td>symbol</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>APOE</td>\n",
       "      <td>11434</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>44903787</td>\n",
       "      <td>44909396</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>rs</td>\n",
       "      <td>1467532231</td>\n",
       "      <td>19</td>\n",
       "      <td>45406049</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Apoe</td>\n",
       "      <td>APOE</td>\n",
       "      <td>symbol</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>APOE</td>\n",
       "      <td>11434</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>44903787</td>\n",
       "      <td>44909396</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>rs</td>\n",
       "      <td>2513533110</td>\n",
       "      <td>19</td>\n",
       "      <td>45406050</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Apoe</td>\n",
       "      <td>APOE</td>\n",
       "      <td>symbol</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>APOE</td>\n",
       "      <td>11434</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>44903787</td>\n",
       "      <td>44909396</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>rs</td>\n",
       "      <td>2513533115</td>\n",
       "      <td>19</td>\n",
       "      <td>45406051</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Apoe</td>\n",
       "      <td>APOE</td>\n",
       "      <td>symbol</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>APOE</td>\n",
       "      <td>11434</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>44903787</td>\n",
       "      <td>44909396</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>rs</td>\n",
       "      <td>955657742</td>\n",
       "      <td>19</td>\n",
       "      <td>45406052</td>\n",
       "      <td>C</td>\n",
       "      <td>A/G/T</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Apoe</td>\n",
       "      <td>APOE</td>\n",
       "      <td>symbol</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>APOE</td>\n",
       "      <td>11434</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>44903787</td>\n",
       "      <td>44909396</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>rs</td>\n",
       "      <td>2513533119</td>\n",
       "      <td>19</td>\n",
       "      <td>45406054</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows √ó 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Input Gene Matched Name Alias Type Alias Source HGNC Symbol  Gene Entity ID  \\\n",
       "0       Apoe         APOE     symbol         HGNC        APOE           11434   \n",
       "1       Apoe         APOE     symbol         HGNC        APOE           11434   \n",
       "2       Apoe         APOE     symbol         HGNC        APOE           11434   \n",
       "3       Apoe         APOE     symbol         HGNC        APOE           11434   \n",
       "4       Apoe         APOE     symbol         HGNC        APOE           11434   \n",
       "\n",
       "  Gene Build Gene Chr (23:X/24:Y) Gene Start (Build 38) Gene End (Build 38)  \\\n",
       "0         38                   19              44903787            44909396   \n",
       "1         38                   19              44903787            44909396   \n",
       "2         38                   19              44903787            44909396   \n",
       "3         38                   19              44903787            44909396   \n",
       "4         38                   19              44903787            44909396   \n",
       "\n",
       "   ... Region Label Variant Type  Variant ID SNP Chr (23:X/24:Y)  \\\n",
       "0  ...         None           rs  1467532231                  19   \n",
       "1  ...         None           rs  2513533110                  19   \n",
       "2  ...         None           rs  2513533115                  19   \n",
       "3  ...         None           rs   955657742                  19   \n",
       "4  ...         None           rs  2513533119                  19   \n",
       "\n",
       "  SNP Pos (Build 37) Ref Allele Alt Allele   DataSource SourceSystem  Note  \n",
       "0           45406049          A          G  dbsnp_chr19         NCBI  <NA>  \n",
       "1           45406050          A          G  dbsnp_chr19         NCBI  <NA>  \n",
       "2           45406051          C          T  dbsnp_chr19         NCBI  <NA>  \n",
       "3           45406052          C      A/G/T  dbsnp_chr19         NCBI  <NA>  \n",
       "4           45406054          A          G  dbsnp_chr19         NCBI  <NA>  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Return Variants in Build 37\n",
    "df = bf.report.run(\n",
    "    \"gene_to_snp\",\n",
    "    input_data = ['Tp53', 'Apoe'],\n",
    "    assembly = 37\n",
    ")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a75d517e-53c1-416e-8472-97b51b3c84fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Input Gene</th>\n",
       "      <th>Matched Name</th>\n",
       "      <th>Alias Type</th>\n",
       "      <th>Alias Source</th>\n",
       "      <th>HGNC Symbol</th>\n",
       "      <th>Gene Entity ID</th>\n",
       "      <th>Gene Build</th>\n",
       "      <th>Gene Chr (23:X/24:Y)</th>\n",
       "      <th>Gene Start (Build 38)</th>\n",
       "      <th>Gene End (Build 38)</th>\n",
       "      <th>...</th>\n",
       "      <th>Region Label</th>\n",
       "      <th>rsID</th>\n",
       "      <th>SNP Chr (23:X/24:Y)</th>\n",
       "      <th>SNP Pos (Build 38)</th>\n",
       "      <th>SNP Pos (Build 37)</th>\n",
       "      <th>Ref Allele</th>\n",
       "      <th>Alt Allele</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>SourceSystem</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ad2</td>\n",
       "      <td>AD2</td>\n",
       "      <td>prev_symbol</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>APOE</td>\n",
       "      <td>11434</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>44903787</td>\n",
       "      <td>44909396</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>7412</td>\n",
       "      <td>19</td>\n",
       "      <td>44908822</td>\n",
       "      <td>45412079</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ad2</td>\n",
       "      <td>AD2</td>\n",
       "      <td>prev_symbol</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>APOE</td>\n",
       "      <td>11434</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>44903787</td>\n",
       "      <td>44909396</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>7256173</td>\n",
       "      <td>19</td>\n",
       "      <td>44904398</td>\n",
       "      <td>45407655</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ad2</td>\n",
       "      <td>AD2</td>\n",
       "      <td>prev_symbol</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>APOE</td>\n",
       "      <td>11434</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>44903787</td>\n",
       "      <td>44909396</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>7259620</td>\n",
       "      <td>19</td>\n",
       "      <td>44904531</td>\n",
       "      <td>45407788</td>\n",
       "      <td>G</td>\n",
       "      <td>A/C</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ad2</td>\n",
       "      <td>AD2</td>\n",
       "      <td>prev_symbol</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>APOE</td>\n",
       "      <td>11434</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>44903787</td>\n",
       "      <td>44909396</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>9282609</td>\n",
       "      <td>19</td>\n",
       "      <td>44905856</td>\n",
       "      <td>45409113</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ad2</td>\n",
       "      <td>AD2</td>\n",
       "      <td>prev_symbol</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>APOE</td>\n",
       "      <td>11434</td>\n",
       "      <td>38</td>\n",
       "      <td>19</td>\n",
       "      <td>44903787</td>\n",
       "      <td>44909396</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>10418135</td>\n",
       "      <td>19</td>\n",
       "      <td>44903541</td>\n",
       "      <td>45406798</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>dbsnp_chr19</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows √ó 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Input Gene Matched Name   Alias Type Alias Source HGNC Symbol  \\\n",
       "0        ad2          AD2  prev_symbol         HGNC        APOE   \n",
       "1        ad2          AD2  prev_symbol         HGNC        APOE   \n",
       "2        ad2          AD2  prev_symbol         HGNC        APOE   \n",
       "3        ad2          AD2  prev_symbol         HGNC        APOE   \n",
       "4        ad2          AD2  prev_symbol         HGNC        APOE   \n",
       "\n",
       "   Gene Entity ID Gene Build Gene Chr (23:X/24:Y) Gene Start (Build 38)  \\\n",
       "0           11434         38                   19              44903787   \n",
       "1           11434         38                   19              44903787   \n",
       "2           11434         38                   19              44903787   \n",
       "3           11434         38                   19              44903787   \n",
       "4           11434         38                   19              44903787   \n",
       "\n",
       "  Gene End (Build 38)  ... Region Label      rsID SNP Chr (23:X/24:Y)  \\\n",
       "0            44909396  ...         None      7412                  19   \n",
       "1            44909396  ...         None   7256173                  19   \n",
       "2            44909396  ...         None   7259620                  19   \n",
       "3            44909396  ...         None   9282609                  19   \n",
       "4            44909396  ...         None  10418135                  19   \n",
       "\n",
       "  SNP Pos (Build 38) SNP Pos (Build 37) Ref Allele Alt Allele   DataSource  \\\n",
       "0           44908822           45412079          C          T  dbsnp_chr19   \n",
       "1           44904398           45407655          C          T  dbsnp_chr19   \n",
       "2           44904531           45407788          G        A/C  dbsnp_chr19   \n",
       "3           44905856           45409113          C          T  dbsnp_chr19   \n",
       "4           44903541           45406798          G          C  dbsnp_chr19   \n",
       "\n",
       "  SourceSystem  Note  \n",
       "0         NCBI  None  \n",
       "1         NCBI  None  \n",
       "2         NCBI  None  \n",
       "3         NCBI  None  \n",
       "4         NCBI  None  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Return Variants in Build 37\n",
    "df = bf.report.run(\"gene_to_snp\", input_data = ['ad2'])\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99540dd4-671e-43fa-8193-abc7e2ba3d47",
   "metadata": {},
   "source": [
    "-------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ad053f2-ca8c-4039-9843-e9ea9662f670",
   "metadata": {},
   "source": [
    "### 4. Run with optional filters / parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9fbae41d-04f7-4546-aaa6-86d26fd41776",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HGNC Symbol</th>\n",
       "      <th>SNP Chr (23:X/24:Y)</th>\n",
       "      <th>SNP Pos (Build 38)</th>\n",
       "      <th>Ref Allele</th>\n",
       "      <th>Alt Allele</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PSEN2</td>\n",
       "      <td>1</td>\n",
       "      <td>226870184</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PSEN2</td>\n",
       "      <td>1</td>\n",
       "      <td>226870192</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PSEN2</td>\n",
       "      <td>1</td>\n",
       "      <td>226870195</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PSEN2</td>\n",
       "      <td>1</td>\n",
       "      <td>226870202</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PSEN2</td>\n",
       "      <td>1</td>\n",
       "      <td>226870205</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  HGNC Symbol SNP Chr (23:X/24:Y) SNP Pos (Build 38) Ref Allele Alt Allele  \\\n",
       "0       PSEN2                   1          226870184          A          C   \n",
       "1       PSEN2                   1          226870192          G          T   \n",
       "2       PSEN2                   1          226870195          G          A   \n",
       "3       PSEN2                   1          226870202          T          A   \n",
       "4       PSEN2                   1          226870205          T          A   \n",
       "\n",
       "   Note  \n",
       "0  <NA>  \n",
       "1  <NA>  \n",
       "2  <NA>  \n",
       "3  <NA>  \n",
       "4  <NA>  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = bf.report.run(\n",
    "    \"gene_to_snp\",\n",
    "    input_data=[\"ENSG00000143801\"],\n",
    "    window_bp=0,        # optional (default: 1000)\n",
    "    output_columns=[\n",
    "        \"HGNC Symbol\",\n",
    "        # \"rsID\",\n",
    "        \"SNP Chr (23:X/24:Y)\",\n",
    "        \"SNP Pos (Build 38)\",\n",
    "        \"Ref Allele\",\n",
    "        \"Alt Allele\", \n",
    "    ],  # optional\n",
    ")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "36fa03d0-5548-4d00-980e-65ede20d8224",
   "metadata": {},
   "outputs": [],
   "source": [
    "# If you need to remember the columns names available here, just type:\n",
    "bf.report.available_columns(\"gene_to_snp\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
