{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0fe43894-4856-4548-bda8-1bc67dca8e8a",
   "metadata": {},
   "source": [
    "<h1>üìò Biofilter Schema Explorer ‚Äî Quick Reference Guide</h1>\n",
    "\n",
    "The Schema Class is an interactive, notebook-friendly interface for exploring all database models, tables, columns, primary keys, and ORM relationships inside Biofilter.\n",
    "\n",
    "Methods:\n",
    "* schema()\n",
    "* schema.table()\n",
    "* schema.overview()\n",
    "* schema.search()\n",
    "* schema.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e66c31f-bc17-4533-b952-713d267b6e59",
   "metadata": {},
   "source": [
    "<h4>Models vs. Tables</h4>\n",
    "\n",
    "In Biofilter3R, a model is a Python ORM class that represents a biological concept (such as a gene, variant, or disease) and defines its columns, relationships, and behavior inside the system. A table, on the other hand, is the physical structure stored in the database where the actual data lives. Every model maps to one table, but models provide additional structure‚Äîsuch as relationships, typing, and a user-friendly API‚Äîwhile tables provide efficient storage and indexing. In practice, you query models when working in Python (e.g., select(GeneMaster)), and you reference tables when executing raw SQL (SELECT * FROM gene_master). The distinction allows Biofilter3R to offer both a high-level, intuitive interface for exploration and a low-level interface for performance and debugging."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8d311e3-b010-419d-840f-3bf68cfc342d",
   "metadata": {},
   "source": [
    "<h3>Start Biofilter3R</h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d58ce367-5ddb-43f0-a86d-44289aca2e23",
   "metadata": {},
   "outputs": [],
   "source": [
    "from biofilter import Biofilter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1f8471f0-d025-4076-974f-28ecea5b7210",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[INFO] ‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê\n",
      "[INFO] üöÄ Initializing Biofilter3R\n",
      "[INFO]    ‚Ä¢ Version: 3.2.0\n",
      "[INFO]    ‚Ä¢ Debug mode: False\n",
      "[INFO]    ‚Ä¢ Config: /home/bioadmin/biofilter/.biofilter.toml\n",
      "[INFO] ‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê\n",
      "[INFO] üîå Database connection established\n",
      "[INFO]    ‚Ä¢ Engine: postgresql+psycopg2\n",
      "[INFO]    ‚Ä¢ Host:   localhost\n",
      "[INFO]    ‚Ä¢ DB:     biofilter\n",
      "[INFO]    ‚Ä¢ Time:   1.0 ms\n",
      "[INFO] ‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê\n"
     ]
    }
   ],
   "source": [
    "# Instance of Biofilter\n",
    "bf = Biofilter()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7a272ff-e441-4393-a616-912314f501db",
   "metadata": {},
   "source": [
    "<h3>1. List all models</h3>\n",
    "\n",
    "Returns a list of all SQLAlchemy models registered inside Biofilter3R.\n",
    "This serves as a quick index of the entire schema."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7e1eb3c0-4fe3-431d-ae27-5b1a23bfbb0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['SystemConfig',\n",
       " 'BiofilterMetadata',\n",
       " 'GenomeAssembly',\n",
       " 'Entity',\n",
       " 'EntityAlias',\n",
       " 'EntityGroup',\n",
       " 'EntityRelationshipType',\n",
       " 'EntityRelationship',\n",
       " 'EntityLocation',\n",
       " 'CurationConflict',\n",
       " 'ConflictStatus',\n",
       " 'ConflictResolution',\n",
       " 'OmicStatus',\n",
       " 'ETLSourceSystem',\n",
       " 'ETLDataSource',\n",
       " 'ETLPackage',\n",
       " 'GeneMaster',\n",
       " 'GeneGroup',\n",
       " 'GeneLocusGroup',\n",
       " 'GeneLocusType',\n",
       " 'GeneGroupMembership',\n",
       " 'VariantSNP',\n",
       " 'VariantSNPMerge',\n",
       " 'VariantGWAS',\n",
       " 'VariantGWASSNP',\n",
       " 'ProteinMaster',\n",
       " 'ProteinPfam',\n",
       " 'ProteinPfamLink',\n",
       " 'ProteinEntity',\n",
       " 'PathwayMaster',\n",
       " 'GOMaster',\n",
       " 'GORelation',\n",
       " 'DiseaseGroup',\n",
       " 'DiseaseGroupMembership',\n",
       " 'DiseaseMaster',\n",
       " 'ChemicalMaster']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bf.schema()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7a16e91-112a-48b4-b443-5c93f1664f71",
   "metadata": {},
   "source": [
    "<h3>2. Filter models by keyword</h3>\n",
    "\n",
    "Returns only the models whose name or group contains \"Gene\"\n",
    "(e.g., GeneMaster, GeneGroup, GeneLocusGroup, etc.).\n",
    "Useful for quickly locating all models related to a biological category."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5fca16ad-f0ee-45ba-8e87-3104929ca57f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GeneMaster',\n",
       " 'GeneGroup',\n",
       " 'GeneLocusGroup',\n",
       " 'GeneLocusType',\n",
       " 'GeneGroupMembership']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bf.schema(\"Gene\") # or bf.schema(keyword=\"Gene\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d0fe7fa-2dcf-4144-9ba9-bf8bcd300959",
   "metadata": {},
   "source": [
    "<h3>3. List all SQL tables</h3>\n",
    "Returns a list of (model_name, table_name) pairs, allowing you to map ORM classes to actual database tables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "cd74a29b-ef31-4199-ab2e-3906a6d230b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('SystemConfig', 'system_config'),\n",
       " ('BiofilterMetadata', 'biofilter_metadata'),\n",
       " ('GenomeAssembly', 'genome_assemblies'),\n",
       " ('Entity', 'entities'),\n",
       " ('EntityAlias', 'entity_aliases'),\n",
       " ('EntityGroup', 'entity_groups'),\n",
       " ('EntityRelationshipType', 'entity_relationship_types'),\n",
       " ('EntityRelationship', 'entity_relationships'),\n",
       " ('EntityLocation', 'entity_locations'),\n",
       " ('CurationConflict', 'curation_conflicts'),\n",
       " ('OmicStatus', 'omic_status'),\n",
       " ('ETLSourceSystem', 'etl_source_systems'),\n",
       " ('ETLDataSource', 'etl_data_sources'),\n",
       " ('ETLPackage', 'etl_packages'),\n",
       " ('GeneMaster', 'gene_masters'),\n",
       " ('GeneGroup', 'gene_groups'),\n",
       " ('GeneLocusGroup', 'gene_locus_groups'),\n",
       " ('GeneLocusType', 'gene_locus_types'),\n",
       " ('GeneGroupMembership', 'gene_group_memberships'),\n",
       " ('VariantSNP', 'variant_snps'),\n",
       " ('VariantSNPMerge', 'variant_snp_merges'),\n",
       " ('VariantGWAS', 'variant_gwas'),\n",
       " ('VariantGWASSNP', 'variant_gwas_snp'),\n",
       " ('ProteinMaster', 'protein_masters'),\n",
       " ('ProteinPfam', 'protein_pfams'),\n",
       " ('ProteinPfamLink', 'protein_pfam_links'),\n",
       " ('ProteinEntity', 'protein_entities'),\n",
       " ('PathwayMaster', 'pathway_masters'),\n",
       " ('GOMaster', 'go_masters'),\n",
       " ('GORelation', 'go_relations'),\n",
       " ('DiseaseGroup', 'disease_groups'),\n",
       " ('DiseaseGroupMembership', 'disease_group_memberships'),\n",
       " ('DiseaseMaster', 'disease_masters'),\n",
       " ('ChemicalMaster', 'chemical_masters')]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bf.schema.tables()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b72c0d4-f4be-4bb3-974f-e03e6e1e17ec",
   "metadata": {},
   "source": [
    "<h3>4. Full schema overview (DataFrame)</h3>\n",
    "Returns a Pandas DataFrame summarizing all Biofilter models, including:\n",
    "\n",
    "* Group (Gene, Variant, Entity, Disease, Protein‚Ä¶)\n",
    "* Model name\n",
    "* Table name\n",
    "* Column names\n",
    "* Primary key columns\n",
    "* Relationship names (SQLAlchemy ORM relationships)\n",
    "\n",
    "This is essentially a fully dynamic ‚Äúdata dictionary‚Äù for the Biofilter database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "42f7b681-4219-4696-8e93-4e7fc964fa26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>model</th>\n",
       "      <th>table</th>\n",
       "      <th>columns</th>\n",
       "      <th>primary_keys</th>\n",
       "      <th>relationships</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>System</td>\n",
       "      <td>SystemConfig</td>\n",
       "      <td>system_config</td>\n",
       "      <td>id, key, value, type, description, editable, c...</td>\n",
       "      <td>id</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>System</td>\n",
       "      <td>BiofilterMetadata</td>\n",
       "      <td>biofilter_metadata</td>\n",
       "      <td>id, schema_version, etl_version, build_hash, d...</td>\n",
       "      <td>id</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>System</td>\n",
       "      <td>GenomeAssembly</td>\n",
       "      <td>genome_assemblies</td>\n",
       "      <td>id, accession, assembly_name, chromosome, crea...</td>\n",
       "      <td>id</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Entity</td>\n",
       "      <td>Entity</td>\n",
       "      <td>entities</td>\n",
       "      <td>id, group_id, has_conflict, is_active, data_so...</td>\n",
       "      <td>id</td>\n",
       "      <td>entity_group, data_source, etl_package, entity...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Entity</td>\n",
       "      <td>EntityAlias</td>\n",
       "      <td>entity_aliases</td>\n",
       "      <td>id, entity_id, group_id, alias_value, alias_ty...</td>\n",
       "      <td>id</td>\n",
       "      <td>entity, entity_group, data_source, etl_package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Entity</td>\n",
       "      <td>EntityGroup</td>\n",
       "      <td>entity_groups</td>\n",
       "      <td>id, name, description</td>\n",
       "      <td>id</td>\n",
       "      <td>entities</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Entity</td>\n",
       "      <td>EntityRelationshipType</td>\n",
       "      <td>entity_relationship_types</td>\n",
       "      <td>id, code, description</td>\n",
       "      <td>id</td>\n",
       "      <td>relationships</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Entity</td>\n",
       "      <td>EntityRelationship</td>\n",
       "      <td>entity_relationships</td>\n",
       "      <td>id, entity_1_id, entity_1_group_id, entity_2_i...</td>\n",
       "      <td>id</td>\n",
       "      <td>entity_1, entity_1_group, entity_2, entity_2_g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Entity</td>\n",
       "      <td>EntityLocation</td>\n",
       "      <td>entity_locations</td>\n",
       "      <td>id, entity_id, entity_group_id, assembly_id, b...</td>\n",
       "      <td>id</td>\n",
       "      <td>entity, entity_group, assembly, data_source, e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Curation</td>\n",
       "      <td>CurationConflict</td>\n",
       "      <td>curation_conflicts</td>\n",
       "      <td>id, data_source_id, entity_type, entity_id, id...</td>\n",
       "      <td>id</td>\n",
       "      <td>data_source</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Curation</td>\n",
       "      <td>OmicStatus</td>\n",
       "      <td>omic_status</td>\n",
       "      <td>id, name, description</td>\n",
       "      <td>id</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ETL</td>\n",
       "      <td>ETLSourceSystem</td>\n",
       "      <td>etl_source_systems</td>\n",
       "      <td>id, name, description, homepage, active, creat...</td>\n",
       "      <td>id</td>\n",
       "      <td>data_sources</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ETL</td>\n",
       "      <td>ETLDataSource</td>\n",
       "      <td>etl_data_sources</td>\n",
       "      <td>id, name, dtp_version, schema_version, source_...</td>\n",
       "      <td>id</td>\n",
       "      <td>source_system, etl_packages</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ETL</td>\n",
       "      <td>ETLPackage</td>\n",
       "      <td>etl_packages</td>\n",
       "      <td>id, data_source_id, status, operation_type, ve...</td>\n",
       "      <td>id</td>\n",
       "      <td>data_source</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Gene</td>\n",
       "      <td>GeneMaster</td>\n",
       "      <td>gene_masters</td>\n",
       "      <td>id, symbol, hgnc_status, omic_status_id, entit...</td>\n",
       "      <td>id</td>\n",
       "      <td>omic_status, entity, data_source, etl_package,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Gene</td>\n",
       "      <td>GeneGroup</td>\n",
       "      <td>gene_groups</td>\n",
       "      <td>id, name, description, data_source_id, etl_pac...</td>\n",
       "      <td>id</td>\n",
       "      <td>data_source, etl_package, memberships</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Gene</td>\n",
       "      <td>GeneLocusGroup</td>\n",
       "      <td>gene_locus_groups</td>\n",
       "      <td>id, name, description, data_source_id, etl_pac...</td>\n",
       "      <td>id</td>\n",
       "      <td>data_source, etl_package, genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Gene</td>\n",
       "      <td>GeneLocusType</td>\n",
       "      <td>gene_locus_types</td>\n",
       "      <td>id, name, description, data_source_id, etl_pac...</td>\n",
       "      <td>id</td>\n",
       "      <td>data_source, etl_package, genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Gene</td>\n",
       "      <td>GeneGroupMembership</td>\n",
       "      <td>gene_group_memberships</td>\n",
       "      <td>gene_id, group_id, data_source_id, etl_package_id</td>\n",
       "      <td>gene_id, group_id</td>\n",
       "      <td>gene, group, data_source, etl_package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Variant</td>\n",
       "      <td>VariantSNP</td>\n",
       "      <td>variant_snps</td>\n",
       "      <td>rs_id, chromosome, position_37, position_38, r...</td>\n",
       "      <td>rs_id</td>\n",
       "      <td>data_source, etl_package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Variant</td>\n",
       "      <td>VariantSNPMerge</td>\n",
       "      <td>variant_snp_merges</td>\n",
       "      <td>rs_obsolete_id, rs_canonical_id, data_source_i...</td>\n",
       "      <td>rs_obsolete_id, rs_canonical_id</td>\n",
       "      <td>data_source, etl_package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Variant</td>\n",
       "      <td>VariantGWAS</td>\n",
       "      <td>variant_gwas</td>\n",
       "      <td>id, pubmed_id, raw_trait, mapped_trait, mapped...</td>\n",
       "      <td>id</td>\n",
       "      <td>data_source, etl_package, snp_links</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Variant</td>\n",
       "      <td>VariantGWASSNP</td>\n",
       "      <td>variant_gwas_snp</td>\n",
       "      <td>id, variant_gwas_id, snp_id, snp_label, snp_rank</td>\n",
       "      <td>id</td>\n",
       "      <td>variant_gwas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Protein</td>\n",
       "      <td>ProteinMaster</td>\n",
       "      <td>protein_masters</td>\n",
       "      <td>id, protein_id, function, location, tissue_exp...</td>\n",
       "      <td>id</td>\n",
       "      <td>data_source, etl_package, pfam_links, protein_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Protein</td>\n",
       "      <td>ProteinPfam</td>\n",
       "      <td>protein_pfams</td>\n",
       "      <td>id, pfam_acc, pfam_id, description, long_descr...</td>\n",
       "      <td>id</td>\n",
       "      <td>data_source, etl_package, protein_links</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Protein</td>\n",
       "      <td>ProteinPfamLink</td>\n",
       "      <td>protein_pfam_links</td>\n",
       "      <td>protein_id, pfam_pk_id, data_source_id, etl_pa...</td>\n",
       "      <td>protein_id, pfam_pk_id</td>\n",
       "      <td>protein_master, pfam, data_source, etl_package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Protein</td>\n",
       "      <td>ProteinEntity</td>\n",
       "      <td>protein_entities</td>\n",
       "      <td>id, entity_id, protein_id, is_isoform, isoform...</td>\n",
       "      <td>id</td>\n",
       "      <td>entity, protein_master, data_source, etl_package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Pathway</td>\n",
       "      <td>PathwayMaster</td>\n",
       "      <td>pathway_masters</td>\n",
       "      <td>id, pathway_id, description, entity_id, data_s...</td>\n",
       "      <td>id</td>\n",
       "      <td>entity, data_source, etl_package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>GO</td>\n",
       "      <td>GOMaster</td>\n",
       "      <td>go_masters</td>\n",
       "      <td>id, go_id, name, namespace, entity_id, data_so...</td>\n",
       "      <td>id</td>\n",
       "      <td>entity, data_source, etl_package, parents, chi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>GO</td>\n",
       "      <td>GORelation</td>\n",
       "      <td>go_relations</td>\n",
       "      <td>id, parent_id, child_id, relation_type, data_s...</td>\n",
       "      <td>id</td>\n",
       "      <td>parent_term, child_term, data_source, etl_package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Disease</td>\n",
       "      <td>DiseaseGroup</td>\n",
       "      <td>disease_groups</td>\n",
       "      <td>id, name, description, data_source_id, etl_pac...</td>\n",
       "      <td>id</td>\n",
       "      <td>data_source, etl_package, memberships</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Disease</td>\n",
       "      <td>DiseaseGroupMembership</td>\n",
       "      <td>disease_group_memberships</td>\n",
       "      <td>id, disease_id, group_id, data_source_id, etl_...</td>\n",
       "      <td>id</td>\n",
       "      <td>disease, group, data_source, etl_package</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Disease</td>\n",
       "      <td>DiseaseMaster</td>\n",
       "      <td>disease_masters</td>\n",
       "      <td>id, disease_id, label, description, omic_statu...</td>\n",
       "      <td>id</td>\n",
       "      <td>omic_status, entity, data_source, etl_package,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Chemical</td>\n",
       "      <td>ChemicalMaster</td>\n",
       "      <td>chemical_masters</td>\n",
       "      <td>id, chemical_id, name, definition, formula, ch...</td>\n",
       "      <td>id</td>\n",
       "      <td>omic_status, entity, data_source, etl_package</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       group                   model                      table  \\\n",
       "0     System            SystemConfig              system_config   \n",
       "1     System       BiofilterMetadata         biofilter_metadata   \n",
       "2     System          GenomeAssembly          genome_assemblies   \n",
       "3     Entity                  Entity                   entities   \n",
       "4     Entity             EntityAlias             entity_aliases   \n",
       "5     Entity             EntityGroup              entity_groups   \n",
       "6     Entity  EntityRelationshipType  entity_relationship_types   \n",
       "7     Entity      EntityRelationship       entity_relationships   \n",
       "8     Entity          EntityLocation           entity_locations   \n",
       "9   Curation        CurationConflict         curation_conflicts   \n",
       "10  Curation              OmicStatus                omic_status   \n",
       "11       ETL         ETLSourceSystem         etl_source_systems   \n",
       "12       ETL           ETLDataSource           etl_data_sources   \n",
       "13       ETL              ETLPackage               etl_packages   \n",
       "14      Gene              GeneMaster               gene_masters   \n",
       "15      Gene               GeneGroup                gene_groups   \n",
       "16      Gene          GeneLocusGroup          gene_locus_groups   \n",
       "17      Gene           GeneLocusType           gene_locus_types   \n",
       "18      Gene     GeneGroupMembership     gene_group_memberships   \n",
       "19   Variant              VariantSNP               variant_snps   \n",
       "20   Variant         VariantSNPMerge         variant_snp_merges   \n",
       "21   Variant             VariantGWAS               variant_gwas   \n",
       "22   Variant          VariantGWASSNP           variant_gwas_snp   \n",
       "23   Protein           ProteinMaster            protein_masters   \n",
       "24   Protein             ProteinPfam              protein_pfams   \n",
       "25   Protein         ProteinPfamLink         protein_pfam_links   \n",
       "26   Protein           ProteinEntity           protein_entities   \n",
       "27   Pathway           PathwayMaster            pathway_masters   \n",
       "28        GO                GOMaster                 go_masters   \n",
       "29        GO              GORelation               go_relations   \n",
       "30   Disease            DiseaseGroup             disease_groups   \n",
       "31   Disease  DiseaseGroupMembership  disease_group_memberships   \n",
       "32   Disease           DiseaseMaster            disease_masters   \n",
       "33  Chemical          ChemicalMaster           chemical_masters   \n",
       "\n",
       "                                              columns  \\\n",
       "0   id, key, value, type, description, editable, c...   \n",
       "1   id, schema_version, etl_version, build_hash, d...   \n",
       "2   id, accession, assembly_name, chromosome, crea...   \n",
       "3   id, group_id, has_conflict, is_active, data_so...   \n",
       "4   id, entity_id, group_id, alias_value, alias_ty...   \n",
       "5                               id, name, description   \n",
       "6                               id, code, description   \n",
       "7   id, entity_1_id, entity_1_group_id, entity_2_i...   \n",
       "8   id, entity_id, entity_group_id, assembly_id, b...   \n",
       "9   id, data_source_id, entity_type, entity_id, id...   \n",
       "10                              id, name, description   \n",
       "11  id, name, description, homepage, active, creat...   \n",
       "12  id, name, dtp_version, schema_version, source_...   \n",
       "13  id, data_source_id, status, operation_type, ve...   \n",
       "14  id, symbol, hgnc_status, omic_status_id, entit...   \n",
       "15  id, name, description, data_source_id, etl_pac...   \n",
       "16  id, name, description, data_source_id, etl_pac...   \n",
       "17  id, name, description, data_source_id, etl_pac...   \n",
       "18  gene_id, group_id, data_source_id, etl_package_id   \n",
       "19  rs_id, chromosome, position_37, position_38, r...   \n",
       "20  rs_obsolete_id, rs_canonical_id, data_source_i...   \n",
       "21  id, pubmed_id, raw_trait, mapped_trait, mapped...   \n",
       "22   id, variant_gwas_id, snp_id, snp_label, snp_rank   \n",
       "23  id, protein_id, function, location, tissue_exp...   \n",
       "24  id, pfam_acc, pfam_id, description, long_descr...   \n",
       "25  protein_id, pfam_pk_id, data_source_id, etl_pa...   \n",
       "26  id, entity_id, protein_id, is_isoform, isoform...   \n",
       "27  id, pathway_id, description, entity_id, data_s...   \n",
       "28  id, go_id, name, namespace, entity_id, data_so...   \n",
       "29  id, parent_id, child_id, relation_type, data_s...   \n",
       "30  id, name, description, data_source_id, etl_pac...   \n",
       "31  id, disease_id, group_id, data_source_id, etl_...   \n",
       "32  id, disease_id, label, description, omic_statu...   \n",
       "33  id, chemical_id, name, definition, formula, ch...   \n",
       "\n",
       "                       primary_keys  \\\n",
       "0                                id   \n",
       "1                                id   \n",
       "2                                id   \n",
       "3                                id   \n",
       "4                                id   \n",
       "5                                id   \n",
       "6                                id   \n",
       "7                                id   \n",
       "8                                id   \n",
       "9                                id   \n",
       "10                               id   \n",
       "11                               id   \n",
       "12                               id   \n",
       "13                               id   \n",
       "14                               id   \n",
       "15                               id   \n",
       "16                               id   \n",
       "17                               id   \n",
       "18                gene_id, group_id   \n",
       "19                            rs_id   \n",
       "20  rs_obsolete_id, rs_canonical_id   \n",
       "21                               id   \n",
       "22                               id   \n",
       "23                               id   \n",
       "24                               id   \n",
       "25           protein_id, pfam_pk_id   \n",
       "26                               id   \n",
       "27                               id   \n",
       "28                               id   \n",
       "29                               id   \n",
       "30                               id   \n",
       "31                               id   \n",
       "32                               id   \n",
       "33                               id   \n",
       "\n",
       "                                        relationships  \n",
       "0                                                      \n",
       "1                                                      \n",
       "2                                                      \n",
       "3   entity_group, data_source, etl_package, entity...  \n",
       "4      entity, entity_group, data_source, etl_package  \n",
       "5                                            entities  \n",
       "6                                       relationships  \n",
       "7   entity_1, entity_1_group, entity_2, entity_2_g...  \n",
       "8   entity, entity_group, assembly, data_source, e...  \n",
       "9                                         data_source  \n",
       "10                                                     \n",
       "11                                       data_sources  \n",
       "12                        source_system, etl_packages  \n",
       "13                                        data_source  \n",
       "14  omic_status, entity, data_source, etl_package,...  \n",
       "15              data_source, etl_package, memberships  \n",
       "16                    data_source, etl_package, genes  \n",
       "17                    data_source, etl_package, genes  \n",
       "18              gene, group, data_source, etl_package  \n",
       "19                           data_source, etl_package  \n",
       "20                           data_source, etl_package  \n",
       "21                data_source, etl_package, snp_links  \n",
       "22                                       variant_gwas  \n",
       "23  data_source, etl_package, pfam_links, protein_...  \n",
       "24            data_source, etl_package, protein_links  \n",
       "25     protein_master, pfam, data_source, etl_package  \n",
       "26   entity, protein_master, data_source, etl_package  \n",
       "27                   entity, data_source, etl_package  \n",
       "28  entity, data_source, etl_package, parents, chi...  \n",
       "29  parent_term, child_term, data_source, etl_package  \n",
       "30              data_source, etl_package, memberships  \n",
       "31           disease, group, data_source, etl_package  \n",
       "32  omic_status, entity, data_source, etl_package,...  \n",
       "33      omic_status, entity, data_source, etl_package  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bf.schema.overview()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b50a0dec-c838-4fc3-b5a4-be8337611e1a",
   "metadata": {},
   "source": [
    "<h3>5. Search schema contents</h3>\n",
    "\n",
    "Searches through:\n",
    "* model names\n",
    "* table names\n",
    "* column names\n",
    "relationship names\n",
    "\n",
    "and returns only the models that match the keyword.\n",
    "\n",
    "Example use cases:\n",
    "* \"variant\" ‚Üí all models/tables related to variants\n",
    "* \"entity\" ‚Üí entity models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "281c001e-5aa7-4820-bc9d-6f2766dd4c8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>model</th>\n",
       "      <th>table</th>\n",
       "      <th>columns</th>\n",
       "      <th>primary_keys</th>\n",
       "      <th>relationships</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>System</td>\n",
       "      <td>GenomeAssembly</td>\n",
       "      <td>genome_assemblies</td>\n",
       "      <td>id, accession, assembly_name, chromosome, crea...</td>\n",
       "      <td>id</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Entity</td>\n",
       "      <td>EntityLocation</td>\n",
       "      <td>entity_locations</td>\n",
       "      <td>id, entity_id, entity_group_id, assembly_id, b...</td>\n",
       "      <td>id</td>\n",
       "      <td>entity, entity_group, assembly, data_source, e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gene</td>\n",
       "      <td>GeneMaster</td>\n",
       "      <td>gene_masters</td>\n",
       "      <td>id, symbol, hgnc_status, omic_status_id, entit...</td>\n",
       "      <td>id</td>\n",
       "      <td>omic_status, entity, data_source, etl_package,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Variant</td>\n",
       "      <td>VariantSNP</td>\n",
       "      <td>variant_snps</td>\n",
       "      <td>rs_id, chromosome, position_37, position_38, r...</td>\n",
       "      <td>rs_id</td>\n",
       "      <td>data_source, etl_package</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     group           model              table  \\\n",
       "0   System  GenomeAssembly  genome_assemblies   \n",
       "1   Entity  EntityLocation   entity_locations   \n",
       "2     Gene      GeneMaster       gene_masters   \n",
       "3  Variant      VariantSNP       variant_snps   \n",
       "\n",
       "                                             columns primary_keys  \\\n",
       "0  id, accession, assembly_name, chromosome, crea...           id   \n",
       "1  id, entity_id, entity_group_id, assembly_id, b...           id   \n",
       "2  id, symbol, hgnc_status, omic_status_id, entit...           id   \n",
       "3  rs_id, chromosome, position_37, position_38, r...        rs_id   \n",
       "\n",
       "                                       relationships  \n",
       "0                                                     \n",
       "1  entity, entity_group, assembly, data_source, e...  \n",
       "2  omic_status, entity, data_source, etl_package,...  \n",
       "3                           data_source, etl_package  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# bf.schema.search(\"variant\")\n",
    "bf.schema.search(\"chrom\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5165e804-337e-4e40-97af-73813b5759a7",
   "metadata": {},
   "source": [
    "<h3>6. Describe a model</h3>\n",
    "\n",
    "Pretty-prints structured information about a specific model:\n",
    "\n",
    "* Table name\n",
    "* Group\n",
    "* All columns + types\n",
    "* Primary keys\n",
    "* ORM relationships"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "44f4be19-aeea-40bf-b134-3c5b5e657cbd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "üìò Model: GeneMaster\n",
      "üîπ Table: gene_masters\n",
      "üî∏ Group: Gene\n",
      "\n",
      "Columns:\n",
      "  ‚Ä¢ id: INTEGER (PK)\n",
      "  ‚Ä¢ symbol: VARCHAR(64)\n",
      "  ‚Ä¢ hgnc_status: VARCHAR(50)\n",
      "  ‚Ä¢ omic_status_id: INTEGER\n",
      "  ‚Ä¢ entity_id: BIGINT\n",
      "  ‚Ä¢ chromosome: VARCHAR(5)\n",
      "  ‚Ä¢ data_source_id: INTEGER\n",
      "  ‚Ä¢ etl_package_id: INTEGER\n",
      "  ‚Ä¢ locus_group_id: INTEGER\n",
      "  ‚Ä¢ locus_type_id: INTEGER\n",
      "\n",
      "Relationships:\n",
      "  ‚Üí omic_status (MANYTOONE)\n",
      "  ‚Üí entity (MANYTOONE)\n",
      "  ‚Üí data_source (MANYTOONE)\n",
      "  ‚Üí etl_package (MANYTOONE)\n",
      "  ‚Üí gene_locus_group (MANYTOONE)\n",
      "  ‚Üí gene_locus_type (MANYTOONE)\n",
      "  ‚Üí group_memberships (ONETOMANY)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "bf.schema.describe(\"GeneMaster\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
