{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "14e44b13-45db-4e76-a95c-117b16104e38",
   "metadata": {},
   "source": [
    "# üß¨ Example Report: Positions to Genes\n",
    "This notebook demonstrates how to run the Position ‚Üí Genes report using Biofilter3R."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4be6574c-8d10-4676-a1ec-f8b355ceade1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from biofilter import Biofilter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "aa2f193c-eea0-4fac-87be-f9e5c7eb1a1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Instance of Biofilter\n",
    "bf = Biofilter(\"postgresql+psycopg2://bioadmin:bioadmin@localhost/biofilter\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "955dfa64-8711-40f1-93f9-ef00abcdb198",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create an input list\n",
    "position_list = [\n",
    "    (\"Y\", 19568371),\n",
    "    (\"Y\", 9999999999),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6e502508-7776-4be9-8f8b-4fa847d04487",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  ‚úÖ Report successfully generated with 3 rows and 2 unique genes.\n",
      "  ‚ÑπÔ∏è Note:\n",
      "  - Variants without associated genes are annotate in the report.\n",
      "  - Some gene metadata is based on external sources (HGNC, NCBI).\n"
     ]
    }
   ],
   "source": [
    "# Run Report\n",
    "result = bf.report.run_report(\n",
    "    \"report_position_to_gene\",\n",
    "    assembly='GRC38',\n",
    "    input_data=position_list,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "765a000a-a976-4a68-8323-a8faac5d6332",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Input Position</th>\n",
       "      <th>Variant ID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref Allele</th>\n",
       "      <th>Alt Allele</th>\n",
       "      <th>Variant Score</th>\n",
       "      <th>Assembly Accession</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>HGNC Status</th>\n",
       "      <th>Locus Group</th>\n",
       "      <th>Locus Type</th>\n",
       "      <th>Gene Groups</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>SourceSystem</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Y:19568371</td>\n",
       "      <td>rs3900</td>\n",
       "      <td>Y</td>\n",
       "      <td>19568371</td>\n",
       "      <td>19568371</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>8.9</td>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>GRCh38.p14</td>\n",
       "      <td>TXLNGY</td>\n",
       "      <td>38610</td>\n",
       "      <td>Approved</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>None</td>\n",
       "      <td>dbsnp_chry</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Y:19568371</td>\n",
       "      <td>rs3900</td>\n",
       "      <td>Y</td>\n",
       "      <td>19568371</td>\n",
       "      <td>19568371</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>8.9</td>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>GRCh38.p14</td>\n",
       "      <td>LOC105377223</td>\n",
       "      <td>54438</td>\n",
       "      <td>Provisional</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>dbsnp_chry</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Y:9999999999</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>No variant or gene found for this position</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Input Position Variant ID   Chr     Start       End Ref Allele Alt Allele  \\\n",
       "0     Y:19568371     rs3900     Y  19568371  19568371          G          C   \n",
       "1     Y:19568371     rs3900     Y  19568371  19568371          G          C   \n",
       "2   Y:9999999999       None  None      None      None       None       None   \n",
       "\n",
       "   Variant Score Assembly Accession    Assembly   Gene Symbol Gene ID  \\\n",
       "0            8.9       NC_000024.10  GRCh38.p14        TXLNGY   38610   \n",
       "1            8.9       NC_000024.10  GRCh38.p14  LOC105377223   54438   \n",
       "2            NaN               None        None          None    None   \n",
       "\n",
       "   HGNC Status Locus Group  Locus Type Gene Groups  DataSource SourceSystem  \\\n",
       "0     Approved  pseudogene  pseudogene        None  dbsnp_chry         NCBI   \n",
       "1  Provisional       ncRNA     unknown   NCBI Gene  dbsnp_chry         NCBI   \n",
       "2         None        None        None        None        None         None   \n",
       "\n",
       "                                         Note  \n",
       "0                                        None  \n",
       "1                                        None  \n",
       "2  No variant or gene found for this position  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Preview the result\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46c95418-9e8a-4f84-a27c-48a8aca3e9cc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
